{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "92094e60",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "import MDAnalysis as mda\n",
    "from MDAnalysis.tests.datafiles import PSF, DCD, GRO, XTC\n",
    "from MDAnalysis.analysis.distances import distance_array\n",
    "import MDAnalysisTests\n",
    "import Bio\n",
    "import Bio.PDB\n",
    "import pytest\n",
    "import pandas as pd\n",
    "import biopandas\n",
    "from biopandas.pdb import PandasPdb\n",
    "import freesasa\n",
    "import os.path\n",
    "from MDAnalysis import AtomGroup\n",
    "import subprocess\n",
    "pd.options.display.max_columns = 999"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e10f6190",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Universe with 10922 atoms>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\user\\anaconda3\\lib\\site-packages\\MDAnalysis\\coordinates\\PDB.py:429: UserWarning: 1 A^3 CRYST1 record, this is usually a placeholder. Unit cell dimensions will be set to None.\n",
      "  warnings.warn(\"1 A^3 CRYST1 record,\"\n"
     ]
    }
   ],
   "source": [
    "KatG = mda.Universe(\"7ag8.pdb\")\n",
    "print(KatG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "53dd420a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#AA_VOLUME (A^3) ('http://www.imgt.org/IMGTeducation/Aide-memoire/_UK/aminoacids/abbreviation.html')\n",
    "#AA_MW (g/mol) ('https://www.thermofisher.com/uk/en/home/references/ambion-tech-support/rna-tools-and-calculators/proteins-and-amino-acids.html')\n",
    "#AA_hydropathy index ('https://doi.org/10.1016/0022-2836(82)90515-0')\n",
    "#AA_Pi ('https://www.sigmaaldrich.com/life-science/metabolomics/learning-center/amino-acid-reference-chart.html')\n",
    "\n",
    "def aa_volume():\n",
    "    aa_volumes = {'A': 88.6, 'R': 173.4, 'N': 114.1, 'D': 111.1, 'C': 108.5,\n",
    "                 'Q': 143.8, 'E': 138.4, 'G': 60.1, 'H': 153.2, 'I': 166.7,\n",
    "                 'L': 166.7, 'K': 168.6, 'M': 162.9, 'F': 189.9, 'P': 112.7,\n",
    "                 'S': 89.0, 'T': 116.1, 'W': 227.8, 'Y': 193.6, 'V': 140.0}\n",
    "    return aa_volumes\n",
    "\n",
    "\n",
    "def MW():\n",
    "    aa_MW = {'A': 89.1, 'R': 174.2, 'N': 132.1, 'D': 133.1, 'C': 121.1,\n",
    "             'E': 147.1, 'Q': 146.2, 'G':75.1, 'H': 155.2, 'I': 131.2,\n",
    "             'L': 131.2, 'K': 146.2, 'M': 149.2, 'F': 165.2, 'P': 115.1,\n",
    "             'S': 105.1, 'T': 119.1, 'W': 204.2, 'Y': 181.2, 'V': 117.1}\n",
    "    return aa_MW\n",
    "\n",
    "\n",
    "def hydropathy():\n",
    "    aa_hydropathy_index = {'A': 1.8, 'R': -4.5, 'N': -3.5, 'D': -3.5,\n",
    "                           'C': 2.5, 'E': -3.5, 'Q': -3.5, 'G': -0.4,\n",
    "                           'H': -3.2, 'I': 4.5, 'L': 3.8, 'K': -3.9,\n",
    "                           'M': 1.9, 'F': 2.8, 'P': -1.6, 'S': -0.8, \n",
    "                           'T': -0.7, 'W': -0.9, 'Y': -1.3, 'V': 4.2}\n",
    "    return aa_hydropathy_index\n",
    "\n",
    "\n",
    "def Pi():\n",
    "    \n",
    "    aa_Pi = {'A': 6, 'R': 10.76, 'N': 5.41, 'D': 2.77, 'C': 5.07, 'E': 3.22,\n",
    "             'Q': 5.65, 'G': 5.97, 'H': 7.59, 'I': 6.02, 'L': 5.98, 'K': 9.74,\n",
    "             'M': 5.74, 'F': 5.48, 'P': 6.3, 'S': 5.68, 'T': 5.6, 'W': 5.89,\n",
    "             'Y': 5.66, 'V': 5.96}\n",
    "    return aa_Pi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "91b57daf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def d_chem_features():\n",
    "    '''calculates change in biochemical feature for every possible mutation'''\n",
    "    \n",
    "    #loads in aa attributes\n",
    "    volume_dict = aa_volume()\n",
    "    MW_dict = MW()\n",
    "    hydropathy_dict = hydropathy()\n",
    "    Pi_dict = Pi()\n",
    "    \n",
    "    #creates df from attribute function data\n",
    "    df = pd.DataFrame.from_dict([volume_dict, MW_dict, hydropathy_dict, Pi_dict]).T.reset_index(0)\n",
    "    df.rename(columns = {'index':'Amino_acid', 0:'volume', 1:'MW', 2:'hydropathy',\n",
    "                        3:'Pi'}, inplace=True)\n",
    "    \n",
    "    #creates dicts for delta attributes from df\n",
    "    d_volume_dict, d_Pi_dict, d_MW_dict, d_hydropathy_dict = {}, {}, {}, {}\n",
    "    for i in df.index:\n",
    "        for j in df.index:\n",
    "            mutation = df['Amino_acid'][i] + df['Amino_acid'][j]\n",
    "            d_volume_dict[mutation] = df['volume'][i] - df['volume'][j]\n",
    "            d_MW_dict[mutation] = df['MW'][i] - df['MW'][j]\n",
    "            d_hydropathy_dict[mutation] = df['hydropathy'][i] - df['hydropathy'][j]\n",
    "            d_Pi_dict[mutation] = df['Pi'][i] - df['Pi'][j]        \n",
    "    \n",
    "    #creates df from delta dictionaries\n",
    "    d_chem_df = pd.DataFrame.from_dict([d_volume_dict, d_MW_dict, d_hydropathy_dict, d_Pi_dict]).T.reset_index(0)\n",
    "    d_chem_df.rename(columns = {'index':'dAA', 0:'d_volume', 1:'d_MW', 2:'d_hydropathy',\n",
    "                        3:'d_Pi'}, inplace=True)\n",
    "    return d_chem_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b3f963fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dAA</th>\n",
       "      <th>d_volume</th>\n",
       "      <th>d_MW</th>\n",
       "      <th>d_hydropathy</th>\n",
       "      <th>d_Pi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AR</td>\n",
       "      <td>-84.8</td>\n",
       "      <td>-85.1</td>\n",
       "      <td>6.3</td>\n",
       "      <td>-4.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AN</td>\n",
       "      <td>-25.5</td>\n",
       "      <td>-43.0</td>\n",
       "      <td>5.3</td>\n",
       "      <td>0.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AD</td>\n",
       "      <td>-22.5</td>\n",
       "      <td>-44.0</td>\n",
       "      <td>5.3</td>\n",
       "      <td>3.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AC</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>VS</td>\n",
       "      <td>51.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>VT</td>\n",
       "      <td>23.9</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>4.9</td>\n",
       "      <td>0.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>VW</td>\n",
       "      <td>-87.8</td>\n",
       "      <td>-87.1</td>\n",
       "      <td>5.1</td>\n",
       "      <td>0.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>VY</td>\n",
       "      <td>-53.6</td>\n",
       "      <td>-64.1</td>\n",
       "      <td>5.5</td>\n",
       "      <td>0.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>VV</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>400 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    dAA  d_volume  d_MW  d_hydropathy  d_Pi\n",
       "0    AA       0.0   0.0           0.0  0.00\n",
       "1    AR     -84.8 -85.1           6.3 -4.76\n",
       "2    AN     -25.5 -43.0           5.3  0.59\n",
       "3    AD     -22.5 -44.0           5.3  3.23\n",
       "4    AC     -19.9 -32.0          -0.7  0.93\n",
       "..   ..       ...   ...           ...   ...\n",
       "395  VS      51.0  12.0           5.0  0.28\n",
       "396  VT      23.9  -2.0           4.9  0.36\n",
       "397  VW     -87.8 -87.1           5.1  0.07\n",
       "398  VY     -53.6 -64.1           5.5  0.30\n",
       "399  VV       0.0   0.0           0.0  0.00\n",
       "\n",
       "[400 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d_chem_features()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "23a9197c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7AG8_1|Chains\n"
     ]
    }
   ],
   "source": [
    "#extracts KatG seq\n",
    "from Bio import SeqIO\n",
    "for record in SeqIO.parse(\"rcsb_pdb_7AG8.fasta\",\"fasta\"):\n",
    "    print(record.id)\n",
    "records = SeqIO.parse(\"rcsb_pdb_7AG8.fasta\",\"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d63f1ce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "KatG_seq = str(record.seq)\n",
    "aa_list = ['A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fa6c86c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#loop to find identify all KatG mutations \n",
    "resnum = 0\n",
    "KatG_mutations = []\n",
    "\n",
    "for i in KatG_seq:\n",
    "    resnum = resnum + 1\n",
    "    for aa in aa_list:\n",
    "        if aa == i:\n",
    "            print('')\n",
    "        else:\n",
    "            ia, resnuma, aaa = str(i), str(resnum), str(aa)\n",
    "            df_add = (ia+resnuma+aaa)\n",
    "            KatG_mutations.append(df_add)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "11962e2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dataframe with all possible KatG chain A mutations \n",
    "data = {'KatG mutations': KatG_mutations}\n",
    "KatG_mutations_df = pd.DataFrame (data, columns = ['KatG mutations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6aef52fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Residue Chain                                     Centre of mass\n",
      "0        M1     A                                                 []\n",
      "1        P2     A                                                 []\n",
      "2        E3     A                                                 []\n",
      "3        Q4     A                                                 []\n",
      "4        H5     A                                                 []\n",
      "..      ...   ...                                                ...\n",
      "735    R736     A  [101.71184776487075, 134.8153974993157, 136.44...\n",
      "736    F737     A  [104.94849311732212, 129.8637529291616, 138.16...\n",
      "737    D738     A  [99.68625343362409, 132.32714759696879, 139.98...\n",
      "738    V739     A  [98.00821963702562, 131.03244203077924, 136.44...\n",
      "739    R740     A  [101.54049856367341, 128.31209317416, 134.9765...\n",
      "\n",
      "[740 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "#identifies the centre of mass of each residue in chain A\n",
    "resid_no = 0\n",
    "residue_no_A = []\n",
    "residue_CoM_A = []\n",
    "chain_A = []\n",
    "\n",
    "for i in KatG_seq:\n",
    "    resid_no = resid_no + 1\n",
    "    residue_no_A.append(i + str(resid_no))\n",
    "    resid_no_string = str(resid_no)\n",
    "    select_atoms_string = 'resid '+ resid_no_string + ' and segid A'\n",
    "    atom_coordinates = KatG.select_atoms(select_atoms_string)\n",
    "    residue_CoM_A.append(atom_coordinates.center_of_mass())\n",
    "    chain_A.append('A')\n",
    "    \n",
    "data = {'Residue': residue_no_A, 'Chain': chain_A, 'Centre of mass': residue_CoM_A}\n",
    "residue_CoM_A_df = pd.DataFrame (data, columns = ['Residue','Chain','Centre of mass'])\n",
    "print(residue_CoM_A_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "23c23152",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Residue Chain                                     Centre of mass\n",
      "0        M1     B                                                 []\n",
      "1        P2     B                                                 []\n",
      "2        E3     B                                                 []\n",
      "3        Q4     B                                                 []\n",
      "4        H5     B                                                 []\n",
      "..      ...   ...                                                ...\n",
      "735    R736     B  [102.56951717105538, 106.68485590620568, 102.0...\n",
      "736    F737     B  [106.1688799400712, 111.39674823454799, 100.59...\n",
      "737    D738     B  [100.97355567927137, 109.2970553620259, 98.230...\n",
      "738    V739     B  [99.15080305467927, 110.36249033674973, 101.78...\n",
      "739    R740     B  [102.86387110467739, 112.88535360473507, 103.8...\n",
      "\n",
      "[740 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "#identifies the centre of mass of each residue in chain B\n",
    "resid_no = 0\n",
    "residue_no_B = []\n",
    "residue_CoM_B = []\n",
    "chain_B = []\n",
    "\n",
    "for i in KatG_seq:\n",
    "    resid_no = resid_no + 1\n",
    "    residue_no_B.append(i + str(resid_no))\n",
    "    resid_no_string = str(resid_no)\n",
    "    select_atoms_string = 'resid '+ resid_no_string + ' and segid B'\n",
    "    atom_coordinates = KatG.select_atoms(select_atoms_string)\n",
    "    residue_CoM_B.append(atom_coordinates.center_of_mass())\n",
    "    chain_B.append('B')\n",
    "    \n",
    "data = {'Residue': residue_no_B, 'Chain': chain_B, 'Centre of mass': residue_CoM_B}\n",
    "residue_CoM_B_df = pd.DataFrame (data, columns = ['Residue','Chain','Centre of mass'])\n",
    "print(residue_CoM_B_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d785da20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adds the chain A and chain B dataframes together\n",
    "residue_CoM_df = residue_CoM_A_df.append(residue_CoM_B_df, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9780ab06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Residue</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Centre of mass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P2</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E3</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q4</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H5</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1475</th>\n",
       "      <td>R736</td>\n",
       "      <td>B</td>\n",
       "      <td>[102.56951717105538, 106.68485590620568, 102.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1476</th>\n",
       "      <td>F737</td>\n",
       "      <td>B</td>\n",
       "      <td>[106.1688799400712, 111.39674823454799, 100.59...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1477</th>\n",
       "      <td>D738</td>\n",
       "      <td>B</td>\n",
       "      <td>[100.97355567927137, 109.2970553620259, 98.230...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1478</th>\n",
       "      <td>V739</td>\n",
       "      <td>B</td>\n",
       "      <td>[99.15080305467927, 110.36249033674973, 101.78...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1479</th>\n",
       "      <td>R740</td>\n",
       "      <td>B</td>\n",
       "      <td>[102.86387110467739, 112.88535360473507, 103.8...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1480 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Residue Chain                                     Centre of mass\n",
       "0         M1     A                                                 []\n",
       "1         P2     A                                                 []\n",
       "2         E3     A                                                 []\n",
       "3         Q4     A                                                 []\n",
       "4         H5     A                                                 []\n",
       "...      ...   ...                                                ...\n",
       "1475    R736     B  [102.56951717105538, 106.68485590620568, 102.0...\n",
       "1476    F737     B  [106.1688799400712, 111.39674823454799, 100.59...\n",
       "1477    D738     B  [100.97355567927137, 109.2970553620259, 98.230...\n",
       "1478    V739     B  [99.15080305467927, 110.36249033674973, 101.78...\n",
       "1479    R740     B  [102.86387110467739, 112.88535360473507, 103.8...\n",
       "\n",
       "[1480 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "residue_CoM_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9cd6a546",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Centre of mass for Ser315 on protomer A: [137.71549016 110.062874   149.22819452]\n",
      "Centre of mass for Ser315 on protomer B: [140.31562649 129.74415378  92.06069301]\n",
      "Centre of mass for Asp137 on protomer A: [133.16871676 115.19637174 144.94393794]\n",
      "Centre of mass for Asp137 on protomer B: [135.49328217 125.00875238  96.04659678]\n",
      "             Distance from Site 1 A\n",
      "Atom number                        \n",
      "10837                      8.124812\n",
      "10838                      6.996807\n",
      "10839                     10.965349\n",
      "10840                     11.538067\n",
      "10841                      7.232062\n",
      "10842                      6.167269\n",
      "10843                      5.783919\n",
      "10844                      6.675063\n",
      "10845                      4.980348\n",
      "10846                      5.944127\n",
      "10847                      5.700443\n",
      "10848                      5.917667\n",
      "10849                      6.626282\n",
      "10850                      5.762992\n",
      "10851                      8.072774\n",
      "10852                      8.848015\n",
      "10853                      9.906793\n",
      "10854                      9.896144\n",
      "10855                      8.693248\n",
      "10856                     11.165976\n",
      "10857                     11.495721\n",
      "10858                     11.239168\n",
      "10859                     12.364128\n",
      "10860                     12.508451\n",
      "10861                     11.501371\n",
      "10862                     13.351467\n",
      "10863                     13.674286\n",
      "10864                     14.755831\n",
      "10865                     10.712170\n",
      "10866                     11.142325\n",
      "10867                     10.281701\n",
      "10868                      9.158696\n",
      "10869                     12.526978\n",
      "10870                     10.662019\n",
      "10871                      9.864463\n",
      "10872                     10.693005\n",
      "10873                     11.776295\n",
      "10874                     10.414757\n",
      "10875                      7.484955\n",
      "10876                      8.813866\n",
      "10877                     10.694182\n",
      "10878                      9.493768\n",
      "10879                      8.923910\n",
      "\n",
      " New site A:\n",
      " [135.55687286 113.46202071 148.68011475]\n",
      "             Distance from Site 1 B\n",
      "Atom number                        \n",
      "10880                      8.107010\n",
      "10881                     11.718632\n",
      "10882                     11.099186\n",
      "10883                      7.093021\n",
      "10884                      9.197226\n",
      "10885                     10.317572\n",
      "10886                     11.228584\n",
      "10887                     10.841160\n",
      "10888                     12.617138\n",
      "10889                     10.638367\n",
      "10890                      9.983967\n",
      "10891                     10.696844\n",
      "10892                     11.581066\n",
      "10893                     10.555311\n",
      "10894                     11.669039\n",
      "10895                     12.640522\n",
      "10896                     12.483518\n",
      "10897                     11.379090\n",
      "10898                     13.814301\n",
      "10899                     13.414285\n",
      "10900                     14.341164\n",
      "10901                     10.018767\n",
      "10902                     10.003643\n",
      "10903                      8.951995\n",
      "10904                      8.175599\n",
      "10905                     11.202010\n",
      "10906                      8.927318\n",
      "10907                      8.057576\n",
      "10908                      6.731700\n",
      "10909                      5.800159\n",
      "10910                      6.138202\n",
      "10911                      7.231101\n",
      "10912                      4.991659\n",
      "10913                      5.854600\n",
      "10914                      5.654332\n",
      "10915                      5.974844\n",
      "10916                      6.607503\n",
      "10917                      5.980443\n",
      "10918                      9.599086\n",
      "10919                     10.865123\n",
      "10920                      8.938081\n",
      "10921                      7.525064\n",
      "10922                      9.044014\n",
      "\n",
      " New site B:\n",
      " [137.77652387 126.92174901  92.18636649]\n"
     ]
    }
   ],
   "source": [
    "#Identifies distance from expected INH binding sites (Site 1) for the residues in each chain\n",
    "\n",
    "#Find Ser315 centre of mass in protomer A and B\n",
    "Ser315_A = KatG.select_atoms('resid 315 and segid A')\n",
    "Ser315_B = KatG.select_atoms('resid 315 and segid B')\n",
    "Ser315_A_CoM = Ser315_A.center_of_mass()\n",
    "Ser315_B_CoM = Ser315_B.center_of_mass()\n",
    "print('Centre of mass for Ser315 on protomer A:', Ser315_A_CoM)\n",
    "print('Centre of mass for Ser315 on protomer B:', Ser315_B_CoM)\n",
    "\n",
    "#Find Asp137 centre of mass in protomer A and B\n",
    "Asp137_A = KatG.select_atoms('resid 137 and segid A')\n",
    "Asp137_B = KatG.select_atoms('resid 137 and segid B')\n",
    "Asp137_A_CoM = Asp137_A.center_of_mass()\n",
    "Asp137_B_CoM = Asp137_B.center_of_mass()\n",
    "print('Centre of mass for Asp137 on protomer A:', Asp137_A_CoM)\n",
    "print('Centre of mass for Asp137 on protomer B:', Asp137_B_CoM)\n",
    "\n",
    "#Identify Hem atoms in protomer A and B\n",
    "Hem_A = KatG.select_atoms('resid 801 and segid A')\n",
    "Hem_B = KatG.select_atoms('resid 801 and segid B')\n",
    "\n",
    "#'Site1_A' is an initial centre of mass in protomer A used to identify the Hem atoms which are closest\n",
    "#to the suspected Isoniazid binding site\n",
    "Site1_A = KatG.select_atoms('resid 315 and segid A', 'resid 137 and segid A')\n",
    "Site1_A_CoM = Site1_A.center_of_mass()\n",
    "haem_distances_A = []\n",
    "haem_atom_number_A = []\n",
    "\n",
    "#Identifies distance of each Hem atoms from suspected binding site\n",
    "for i in Hem_A:\n",
    "    atom_position = i.position\n",
    "    dist = numpy.linalg.norm(Site1_A_CoM-atom_position)\n",
    "    haem_distances_A.append(dist)\n",
    "    haem_atom_number_A.append(str(i)[6:11])\n",
    "    \n",
    "#Creates dataframe to show atom number and distance\n",
    "data = {'Atom number': haem_atom_number_A, 'Distance from Site 1 A': haem_distances_A}\n",
    "distance_site1_A_df = pd.DataFrame (data, columns = ['Atom number', 'Distance from Site 1 A'])\n",
    "distance_site1_A_df = distance_site1_A_df.set_index('Atom number')\n",
    "print(distance_site1_A_df)\n",
    "\n",
    "#Uses a carbon from each of Hem, Ser315 and Asp137, close to the suspected binding site\n",
    "#to find a new, more accurate, binding site location\n",
    "Site1_A_true = KatG.select_atoms('bynum 10844','bynum 2112','bynum 890')\n",
    "Site1_A_true_CoM = Site1_A_true.center_of_mass()\n",
    "print('\\n','New site A:''\\n', Site1_A_true_CoM)\n",
    "\n",
    "#'Site1_B' is an initial centre of mass in protomer B used to identify the Hem atoms which are closest\n",
    "#to the suspected Isoniazid binding site\n",
    "Site1_B = KatG.select_atoms('resid 315 and segid B', 'resid 137 and segid B')\n",
    "Site1_B_CoM = Site1_B.center_of_mass()\n",
    "haem_distances_B = []\n",
    "haem_atom_number_B = []\n",
    "\n",
    "#Identifies distance of each Hem atoms from suspected binding site\n",
    "for i in Hem_B:\n",
    "    atom_position = i.position\n",
    "    dist = numpy.linalg.norm(Site1_B_CoM-atom_position)\n",
    "    haem_distances_B.append(dist)\n",
    "    haem_atom_number_B.append(str(i)[6:11])\n",
    "    \n",
    "#Creates dataframe to show atom number and distance\n",
    "data = {'Atom number': haem_atom_number_B, 'Distance from Site 1 B': haem_distances_B}\n",
    "distance_site1_B_df = pd.DataFrame (data, columns = ['Atom number', 'Distance from Site 1 B'])\n",
    "distance_site1_B_df = distance_site1_B_df.set_index('Atom number')\n",
    "print(distance_site1_B_df)\n",
    "\n",
    "#Uses a carbon from each of Hem, Ser315 and Asp137, close to the suspected binding site\n",
    "#to find a new, more accurate, binding site location\n",
    "Site1_B_true = KatG.select_atoms('bynum 10911','bynum 7582','bynum 6236')\n",
    "Site1_B_true_CoM = Site1_B_true.center_of_mass()\n",
    "print('\\n','New site B:''\\n', Site1_B_true_CoM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1ef2c98b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Residue</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Centre of mass</th>\n",
       "      <th>Site 1 Coordinates</th>\n",
       "      <th>Distance from Site 1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "      <td>[135.5568728593392, 113.46202070912435, 148.68...</td>\n",
       "      <td>No centre of mass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P2</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "      <td>[135.5568728593392, 113.46202070912435, 148.68...</td>\n",
       "      <td>No centre of mass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E3</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "      <td>[135.5568728593392, 113.46202070912435, 148.68...</td>\n",
       "      <td>No centre of mass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q4</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "      <td>[135.5568728593392, 113.46202070912435, 148.68...</td>\n",
       "      <td>No centre of mass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H5</td>\n",
       "      <td>A</td>\n",
       "      <td>[]</td>\n",
       "      <td>[135.5568728593392, 113.46202070912435, 148.68...</td>\n",
       "      <td>No centre of mass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1475</th>\n",
       "      <td>R736</td>\n",
       "      <td>B</td>\n",
       "      <td>[102.56951717105538, 106.68485590620568, 102.0...</td>\n",
       "      <td>[137.7765238725133, 126.92174901271466, 92.186...</td>\n",
       "      <td>41.195753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1476</th>\n",
       "      <td>F737</td>\n",
       "      <td>B</td>\n",
       "      <td>[106.1688799400712, 111.39674823454799, 100.59...</td>\n",
       "      <td>[137.7765238725133, 126.92174901271466, 92.186...</td>\n",
       "      <td>35.615301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1477</th>\n",
       "      <td>D738</td>\n",
       "      <td>B</td>\n",
       "      <td>[100.97355567927137, 109.2970553620259, 98.230...</td>\n",
       "      <td>[137.7765238725133, 126.92174901271466, 92.186...</td>\n",
       "      <td>40.836819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1478</th>\n",
       "      <td>V739</td>\n",
       "      <td>B</td>\n",
       "      <td>[99.15080305467927, 110.36249033674973, 101.78...</td>\n",
       "      <td>[137.7765238725133, 126.92174901271466, 92.186...</td>\n",
       "      <td>42.519218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1479</th>\n",
       "      <td>R740</td>\n",
       "      <td>B</td>\n",
       "      <td>[102.86387110467739, 112.88535360473507, 103.8...</td>\n",
       "      <td>[137.7765238725133, 126.92174901271466, 92.186...</td>\n",
       "      <td>38.661813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1480 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Residue Chain                                     Centre of mass  \\\n",
       "0         M1     A                                                 []   \n",
       "1         P2     A                                                 []   \n",
       "2         E3     A                                                 []   \n",
       "3         Q4     A                                                 []   \n",
       "4         H5     A                                                 []   \n",
       "...      ...   ...                                                ...   \n",
       "1475    R736     B  [102.56951717105538, 106.68485590620568, 102.0...   \n",
       "1476    F737     B  [106.1688799400712, 111.39674823454799, 100.59...   \n",
       "1477    D738     B  [100.97355567927137, 109.2970553620259, 98.230...   \n",
       "1478    V739     B  [99.15080305467927, 110.36249033674973, 101.78...   \n",
       "1479    R740     B  [102.86387110467739, 112.88535360473507, 103.8...   \n",
       "\n",
       "                                     Site 1 Coordinates Distance from Site 1  \n",
       "0     [135.5568728593392, 113.46202070912435, 148.68...    No centre of mass  \n",
       "1     [135.5568728593392, 113.46202070912435, 148.68...    No centre of mass  \n",
       "2     [135.5568728593392, 113.46202070912435, 148.68...    No centre of mass  \n",
       "3     [135.5568728593392, 113.46202070912435, 148.68...    No centre of mass  \n",
       "4     [135.5568728593392, 113.46202070912435, 148.68...    No centre of mass  \n",
       "...                                                 ...                  ...  \n",
       "1475  [137.7765238725133, 126.92174901271466, 92.186...            41.195753  \n",
       "1476  [137.7765238725133, 126.92174901271466, 92.186...            35.615301  \n",
       "1477  [137.7765238725133, 126.92174901271466, 92.186...            40.836819  \n",
       "1478  [137.7765238725133, 126.92174901271466, 92.186...            42.519218  \n",
       "1479  [137.7765238725133, 126.92174901271466, 92.186...            38.661813  \n",
       "\n",
       "[1480 rows x 5 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#adds a new column for distance from site 1 for each residue in that given chain\n",
    "#Note - Site 1 chain B is done later as only residues in chain A will be used for ML model\n",
    "Site1_coord = []\n",
    "for i in residue_CoM_df['Chain']:\n",
    "    if (i == 'A'):\n",
    "        Site1_coord.append(Site1_A_true_CoM)\n",
    "    else:\n",
    "        Site1_coord.append(Site1_B_true_CoM)\n",
    "residue_CoM_df['Site 1 Coordinates'] = Site1_coord\n",
    "\n",
    "Site1_dist = []\n",
    "Site1_dist_no = -1\n",
    "\n",
    "for i in residue_CoM_df['Centre of mass']:\n",
    "    Site1_dist_no = Site1_dist_no + 1 \n",
    "    chain = residue_CoM_df.loc[Site1_dist_no,'Chain']\n",
    "    CoM = residue_CoM_df.loc[Site1_dist_no,'Centre of mass']\n",
    "    if len(CoM) == 0:\n",
    "        Site1_dist.append('No centre of mass')\n",
    "    else:\n",
    "        if chain == 'A':\n",
    "            dist = numpy.linalg.norm(Site1_A_CoM-i)\n",
    "            Site1_dist.append(dist)\n",
    "        else:\n",
    "            dist = numpy.linalg.norm(Site1_B_CoM-i)\n",
    "            Site1_dist.append(dist)\n",
    "            \n",
    "residue_CoM_df['Distance from Site 1'] = Site1_dist\n",
    "residue_CoM_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3f15caa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#creates a dataframe for all mutations in chain A\n",
    "KatG_mutations_chain_A = []\n",
    "for i in KatG_mutations_df['KatG mutations']:\n",
    "    KatG_mutations_chain_A.append('A')\n",
    "\n",
    "KatG_mutations_df['Chain'] =  KatG_mutations_chain_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0bd71ca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "KatG_mutations_df_B = KatG_mutations_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "10b28585",
   "metadata": {},
   "outputs": [],
   "source": [
    "#creates a dataframe for all mutations in chain B\n",
    "del KatG_mutations_df_B['Chain']\n",
    "KatG_mutations_chain_B = []\n",
    "for i in KatG_mutations_df_B['KatG mutations']:\n",
    "    KatG_mutations_chain_B.append('B')\n",
    "\n",
    "KatG_mutations_df_B['Chain'] =  KatG_mutations_chain_B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7da19e30",
   "metadata": {},
   "outputs": [],
   "source": [
    "#joins chain A and chain B dataframes together\n",
    "KatG_mutations_df = KatG_mutations_df.append(KatG_mutations_df_B, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c836bac2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KatG mutations</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Unique ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1A</td>\n",
       "      <td>A</td>\n",
       "      <td>M1A A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1C</td>\n",
       "      <td>A</td>\n",
       "      <td>M1C A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M1D</td>\n",
       "      <td>A</td>\n",
       "      <td>M1D A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M1E</td>\n",
       "      <td>A</td>\n",
       "      <td>M1E A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1F</td>\n",
       "      <td>A</td>\n",
       "      <td>M1F A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28115</th>\n",
       "      <td>R740S</td>\n",
       "      <td>B</td>\n",
       "      <td>R740S B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28116</th>\n",
       "      <td>R740T</td>\n",
       "      <td>B</td>\n",
       "      <td>R740T B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28117</th>\n",
       "      <td>R740V</td>\n",
       "      <td>B</td>\n",
       "      <td>R740V B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28118</th>\n",
       "      <td>R740W</td>\n",
       "      <td>B</td>\n",
       "      <td>R740W B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28119</th>\n",
       "      <td>R740Y</td>\n",
       "      <td>B</td>\n",
       "      <td>R740Y B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28120 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      KatG mutations Chain Unique ID\n",
       "0                M1A     A     M1A A\n",
       "1                M1C     A     M1C A\n",
       "2                M1D     A     M1D A\n",
       "3                M1E     A     M1E A\n",
       "4                M1F     A     M1F A\n",
       "...              ...   ...       ...\n",
       "28115          R740S     B   R740S B\n",
       "28116          R740T     B   R740T B\n",
       "28117          R740V     B   R740V B\n",
       "28118          R740W     B   R740W B\n",
       "28119          R740Y     B   R740Y B\n",
       "\n",
       "[28120 rows x 3 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#creates unique ID for each possible KatG mutation\n",
    "\n",
    "Unique_ID = []\n",
    "mutation_no = -1\n",
    "\n",
    "for i in KatG_mutations_df['KatG mutations']:\n",
    "    mutation_no = mutation_no + 1\n",
    "    if mutation_no < 14060:\n",
    "        ID = i + ' ' + 'A'\n",
    "    else:\n",
    "        ID = i + ' ' + 'B'\n",
    "    Unique_ID.append(ID)\n",
    "\n",
    "KatG_mutations_df['Unique ID'] =  Unique_ID\n",
    "KatG_mutations_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b460cf18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KatG mutations</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Unique ID</th>\n",
       "      <th>Distance from Site 1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1A</td>\n",
       "      <td>A</td>\n",
       "      <td>M1A A</td>\n",
       "      <td>No centre of mass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1C</td>\n",
       "      <td>A</td>\n",
       "      <td>M1C A</td>\n",
       "      <td>No centre of mass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M1D</td>\n",
       "      <td>A</td>\n",
       "      <td>M1D A</td>\n",
       "      <td>No centre of mass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M1E</td>\n",
       "      <td>A</td>\n",
       "      <td>M1E A</td>\n",
       "      <td>No centre of mass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1F</td>\n",
       "      <td>A</td>\n",
       "      <td>M1F A</td>\n",
       "      <td>No centre of mass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28115</th>\n",
       "      <td>R740S</td>\n",
       "      <td>B</td>\n",
       "      <td>R740S B</td>\n",
       "      <td>38.661813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28116</th>\n",
       "      <td>R740T</td>\n",
       "      <td>B</td>\n",
       "      <td>R740T B</td>\n",
       "      <td>38.661813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28117</th>\n",
       "      <td>R740V</td>\n",
       "      <td>B</td>\n",
       "      <td>R740V B</td>\n",
       "      <td>38.661813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28118</th>\n",
       "      <td>R740W</td>\n",
       "      <td>B</td>\n",
       "      <td>R740W B</td>\n",
       "      <td>38.661813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28119</th>\n",
       "      <td>R740Y</td>\n",
       "      <td>B</td>\n",
       "      <td>R740Y B</td>\n",
       "      <td>38.661813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28120 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      KatG mutations Chain Unique ID Distance from Site 1\n",
       "0                M1A     A     M1A A    No centre of mass\n",
       "1                M1C     A     M1C A    No centre of mass\n",
       "2                M1D     A     M1D A    No centre of mass\n",
       "3                M1E     A     M1E A    No centre of mass\n",
       "4                M1F     A     M1F A    No centre of mass\n",
       "...              ...   ...       ...                  ...\n",
       "28115          R740S     B   R740S B            38.661813\n",
       "28116          R740T     B   R740T B            38.661813\n",
       "28117          R740V     B   R740V B            38.661813\n",
       "28118          R740W     B   R740W B            38.661813\n",
       "28119          R740Y     B   R740Y B            38.661813\n",
       "\n",
       "[28120 rows x 4 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#adds distance from site 1 column\n",
    "\n",
    "KatG_mutations_dist = []\n",
    "mutation_no = -1\n",
    "\n",
    "#cycles through all possible KatG mutations\n",
    "for a in KatG_mutations_df['Unique ID']:\n",
    "    mutation_no = mutation_no + 1\n",
    "    #cycles through all KatG residues looking for a match to the residue being mutated\n",
    "    for b in residue_CoM_df['Residue']:\n",
    "        if a[:-3] == b:\n",
    "            if mutation_no < 14060:\n",
    "                row_no = residue_CoM_df[residue_CoM_df['Residue']==b].index[0]\n",
    "            else:\n",
    "                row_no = residue_CoM_df[residue_CoM_df['Residue']==b].index[1]\n",
    "            Site1_dist = residue_CoM_df.loc[row_no,'Distance from Site 1']\n",
    "    KatG_mutations_dist.append(Site1_dist)\n",
    "    \n",
    "KatG_mutations_df['Distance from Site 1'] =  KatG_mutations_dist\n",
    "KatG_mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5e387a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_chem_features = d_chem_features()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "218881d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KatG mutations</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Unique ID</th>\n",
       "      <th>Distance from Site 1</th>\n",
       "      <th>d_volume</th>\n",
       "      <th>d_MW</th>\n",
       "      <th>d_hydropathy</th>\n",
       "      <th>d_Pi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1A</td>\n",
       "      <td>A</td>\n",
       "      <td>M1A A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>74.3</td>\n",
       "      <td>60.1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>-0.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1C</td>\n",
       "      <td>A</td>\n",
       "      <td>M1C A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>54.4</td>\n",
       "      <td>28.1</td>\n",
       "      <td>-0.6</td>\n",
       "      <td>0.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M1D</td>\n",
       "      <td>A</td>\n",
       "      <td>M1D A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>51.8</td>\n",
       "      <td>16.1</td>\n",
       "      <td>5.4</td>\n",
       "      <td>2.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M1E</td>\n",
       "      <td>A</td>\n",
       "      <td>M1E A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>24.5</td>\n",
       "      <td>2.1</td>\n",
       "      <td>5.4</td>\n",
       "      <td>2.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1F</td>\n",
       "      <td>A</td>\n",
       "      <td>M1F A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>-27.0</td>\n",
       "      <td>-16.0</td>\n",
       "      <td>-0.9</td>\n",
       "      <td>0.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28115</th>\n",
       "      <td>R740S</td>\n",
       "      <td>B</td>\n",
       "      <td>R740S B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>84.4</td>\n",
       "      <td>69.1</td>\n",
       "      <td>-3.7</td>\n",
       "      <td>5.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28116</th>\n",
       "      <td>R740T</td>\n",
       "      <td>B</td>\n",
       "      <td>R740T B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>57.3</td>\n",
       "      <td>55.1</td>\n",
       "      <td>-3.8</td>\n",
       "      <td>5.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28117</th>\n",
       "      <td>R740V</td>\n",
       "      <td>B</td>\n",
       "      <td>R740V B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>33.4</td>\n",
       "      <td>57.1</td>\n",
       "      <td>-8.7</td>\n",
       "      <td>4.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28118</th>\n",
       "      <td>R740W</td>\n",
       "      <td>B</td>\n",
       "      <td>R740W B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>-54.4</td>\n",
       "      <td>-30.0</td>\n",
       "      <td>-3.6</td>\n",
       "      <td>4.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28119</th>\n",
       "      <td>R740Y</td>\n",
       "      <td>B</td>\n",
       "      <td>R740Y B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>-3.2</td>\n",
       "      <td>5.10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28120 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      KatG mutations Chain Unique ID Distance from Site 1  d_volume  d_MW  \\\n",
       "0                M1A     A     M1A A    No centre of mass      74.3  60.1   \n",
       "1                M1C     A     M1C A    No centre of mass      54.4  28.1   \n",
       "2                M1D     A     M1D A    No centre of mass      51.8  16.1   \n",
       "3                M1E     A     M1E A    No centre of mass      24.5   2.1   \n",
       "4                M1F     A     M1F A    No centre of mass     -27.0 -16.0   \n",
       "...              ...   ...       ...                  ...       ...   ...   \n",
       "28115          R740S     B   R740S B            38.661813      84.4  69.1   \n",
       "28116          R740T     B   R740T B            38.661813      57.3  55.1   \n",
       "28117          R740V     B   R740V B            38.661813      33.4  57.1   \n",
       "28118          R740W     B   R740W B            38.661813     -54.4 -30.0   \n",
       "28119          R740Y     B   R740Y B            38.661813     -20.2  -7.0   \n",
       "\n",
       "       d_hydropathy  d_Pi  \n",
       "0               0.1 -0.26  \n",
       "1              -0.6  0.67  \n",
       "2               5.4  2.97  \n",
       "3               5.4  2.52  \n",
       "4              -0.9  0.26  \n",
       "...             ...   ...  \n",
       "28115          -3.7  5.08  \n",
       "28116          -3.8  5.16  \n",
       "28117          -8.7  4.80  \n",
       "28118          -3.6  4.87  \n",
       "28119          -3.2  5.10  \n",
       "\n",
       "[28120 rows x 8 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#add d_chem_features columns to KatG_mutations_df\n",
    "d_volume = []\n",
    "d_MW = []\n",
    "d_hydropathy = []\n",
    "d_Pi = []\n",
    "\n",
    "for a in KatG_mutations_df['KatG mutations']:\n",
    "    res1 = a[0]\n",
    "    res2 = a[-1]\n",
    "    AA = res1 + res2\n",
    "    for b in d_chem_features['dAA']:\n",
    "        if str(AA) == str(b):\n",
    "            row_no = d_chem_features[d_chem_features['dAA']==b].index[0]\n",
    "            vol = d_chem_features.loc[row_no,'d_volume']\n",
    "            d_volume.append(vol)\n",
    "            MW = d_chem_features.loc[row_no,'d_MW']\n",
    "            d_MW.append(MW)\n",
    "            hydrop =d_chem_features.loc[row_no,'d_hydropathy']\n",
    "            d_hydropathy.append(hydrop)\n",
    "            pi =d_chem_features.loc[row_no,'d_Pi']\n",
    "            d_Pi.append(pi)\n",
    "            \n",
    "KatG_mutations_df['d_volume'] = d_volume\n",
    "KatG_mutations_df['d_MW'] = d_MW\n",
    "KatG_mutations_df['d_hydropathy'] = d_hydropathy\n",
    "KatG_mutations_df['d_Pi'] = d_Pi\n",
    "\n",
    "KatG_mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e01f31f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#creates a list of all possible codons\n",
    "codon_list = []\n",
    "bases = ['T','A','C','G']\n",
    "for i in bases:\n",
    "    for h in bases:\n",
    "        for z in bases:\n",
    "            codon = (i,h,z)\n",
    "            codon_list.append(codon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6d1a33b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#function to convert lists to strings\n",
    "def ListToString(s):\n",
    "    str1 = ''\n",
    "    for ele in s:\n",
    "        str1 += ele\n",
    "    return str1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a7fc602a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start codon</th>\n",
       "      <th>End codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4091</th>\n",
       "      <td>GGG</td>\n",
       "      <td>GCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4092</th>\n",
       "      <td>GGG</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4093</th>\n",
       "      <td>GGG</td>\n",
       "      <td>GGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4094</th>\n",
       "      <td>GGG</td>\n",
       "      <td>GGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4095</th>\n",
       "      <td>GGG</td>\n",
       "      <td>GGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4096 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Start codon End codon\n",
       "0            TTT       TTT\n",
       "1            TTT       TTA\n",
       "2            TTT       TTC\n",
       "3            TTT       TTG\n",
       "4            TTT       TAT\n",
       "...          ...       ...\n",
       "4091         GGG       GCG\n",
       "4092         GGG       GGT\n",
       "4093         GGG       GGA\n",
       "4094         GGG       GGC\n",
       "4095         GGG       GGG\n",
       "\n",
       "[4096 rows x 2 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#creates dataframe with all possible codon to codon mutations\n",
    "codon_list_start = []\n",
    "codon_list_end = []\n",
    "for i in codon_list:\n",
    "    for a in codon_list:\n",
    "        start_codon = ListToString(i)\n",
    "        end_codon = ListToString(a)\n",
    "        codon_list_start.append(start_codon)\n",
    "        codon_list_end.append(end_codon)\n",
    "        \n",
    "data = {'Start codon': codon_list_start, 'End codon': codon_list_end}\n",
    "codon_df = pd.DataFrame (data, columns = ['Start codon', 'End codon'])\n",
    "codon_df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6e237500",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    }
   ],
   "source": [
    "#checks if code can determine the number of bases changed in a codon mutation\n",
    "codon_1 = 'TAC'\n",
    "codon_2 = 'GCC'\n",
    "base_no = -1\n",
    "errors = 0\n",
    "for i in codon_1:\n",
    "    base_no = base_no + 1\n",
    "    if codon_2[int(base_no)] == i:\n",
    "        errors += 0\n",
    "    else:\n",
    "        errors += 1\n",
    "        \n",
    "print(errors)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "77ce1efd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start codon</th>\n",
       "      <th>End codon</th>\n",
       "      <th>Number of substitutions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTT</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TAT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4091</th>\n",
       "      <td>GGG</td>\n",
       "      <td>GCG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4092</th>\n",
       "      <td>GGG</td>\n",
       "      <td>GGT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4093</th>\n",
       "      <td>GGG</td>\n",
       "      <td>GGA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4094</th>\n",
       "      <td>GGG</td>\n",
       "      <td>GGC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4095</th>\n",
       "      <td>GGG</td>\n",
       "      <td>GGG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4096 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Start codon End codon  Number of substitutions\n",
       "0            TTT       TTT                        0\n",
       "1            TTT       TTA                        1\n",
       "2            TTT       TTC                        1\n",
       "3            TTT       TTG                        1\n",
       "4            TTT       TAT                        1\n",
       "...          ...       ...                      ...\n",
       "4091         GGG       GCG                        1\n",
       "4092         GGG       GGT                        1\n",
       "4093         GGG       GGA                        1\n",
       "4094         GGG       GGC                        1\n",
       "4095         GGG       GGG                        0\n",
       "\n",
       "[4096 rows x 3 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#adds column for number of substitutions for each codon mutation\n",
    "substitutions = []\n",
    "row_no = -1\n",
    "for i in codon_list_start:\n",
    "    codon_1 = i\n",
    "    row_no = row_no + 1\n",
    "    codon_2 = codon_df.loc[row_no,'End codon']\n",
    "    base_no = -1\n",
    "    subs = 0\n",
    "    for a in codon_1:\n",
    "        base_no = base_no + 1\n",
    "        if codon_2[int(base_no)] == a:\n",
    "            subs += 0\n",
    "        else:\n",
    "            subs += 1\n",
    "    substitutions.append(subs)\n",
    "\n",
    "codon_df['Number of substitutions'] = substitutions \n",
    "codon_df    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "02488867",
   "metadata": {},
   "outputs": [],
   "source": [
    "#loads dataframe showing what residue each codon codes for - but has spaces and additional rows (draft)\n",
    "aa_codons_drft = pd.read_csv(r'C:\\Users\\user\\anaconda3\\envs\\mdaenv\\Inh_resistance\\amino_acid_codons.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "578ae1fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Amino Acid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codon</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TTT</th>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTA</th>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTC</th>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTG</th>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAT</th>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCG</th>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGT</th>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGA</th>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGC</th>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGG</th>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Amino Acid\n",
       "Codon            \n",
       "TTT             F\n",
       "TTA             L\n",
       "TTC             F\n",
       "TTG             L\n",
       "TAT             Y\n",
       "...           ...\n",
       "GCG             A\n",
       "GGT             G\n",
       "GGA             G\n",
       "GGC             G\n",
       "GGG             G\n",
       "\n",
       "[64 rows x 1 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#converts draft codon df to formatted df\n",
    "aa_codons = aa_codons_drft.iloc[:-3,]\n",
    "aa_codons = aa_codons.rename(columns ={0:'Codon',1:'Amino Acid'})\n",
    "aa_codons.set_index('Codon ', inplace = True)\n",
    "aa_codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "064cf00c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KatG mutations</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Unique ID</th>\n",
       "      <th>Distance from Site 1</th>\n",
       "      <th>d_volume</th>\n",
       "      <th>d_MW</th>\n",
       "      <th>d_hydropathy</th>\n",
       "      <th>d_Pi</th>\n",
       "      <th>Start residue</th>\n",
       "      <th>End residue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1A</td>\n",
       "      <td>A</td>\n",
       "      <td>M1A A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>74.3</td>\n",
       "      <td>60.1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>-0.26</td>\n",
       "      <td>M</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1C</td>\n",
       "      <td>A</td>\n",
       "      <td>M1C A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>54.4</td>\n",
       "      <td>28.1</td>\n",
       "      <td>-0.6</td>\n",
       "      <td>0.67</td>\n",
       "      <td>M</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M1D</td>\n",
       "      <td>A</td>\n",
       "      <td>M1D A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>51.8</td>\n",
       "      <td>16.1</td>\n",
       "      <td>5.4</td>\n",
       "      <td>2.97</td>\n",
       "      <td>M</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M1E</td>\n",
       "      <td>A</td>\n",
       "      <td>M1E A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>24.5</td>\n",
       "      <td>2.1</td>\n",
       "      <td>5.4</td>\n",
       "      <td>2.52</td>\n",
       "      <td>M</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1F</td>\n",
       "      <td>A</td>\n",
       "      <td>M1F A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>-27.0</td>\n",
       "      <td>-16.0</td>\n",
       "      <td>-0.9</td>\n",
       "      <td>0.26</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28115</th>\n",
       "      <td>R740S</td>\n",
       "      <td>B</td>\n",
       "      <td>R740S B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>84.4</td>\n",
       "      <td>69.1</td>\n",
       "      <td>-3.7</td>\n",
       "      <td>5.08</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28116</th>\n",
       "      <td>R740T</td>\n",
       "      <td>B</td>\n",
       "      <td>R740T B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>57.3</td>\n",
       "      <td>55.1</td>\n",
       "      <td>-3.8</td>\n",
       "      <td>5.16</td>\n",
       "      <td>R</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28117</th>\n",
       "      <td>R740V</td>\n",
       "      <td>B</td>\n",
       "      <td>R740V B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>33.4</td>\n",
       "      <td>57.1</td>\n",
       "      <td>-8.7</td>\n",
       "      <td>4.80</td>\n",
       "      <td>R</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28118</th>\n",
       "      <td>R740W</td>\n",
       "      <td>B</td>\n",
       "      <td>R740W B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>-54.4</td>\n",
       "      <td>-30.0</td>\n",
       "      <td>-3.6</td>\n",
       "      <td>4.87</td>\n",
       "      <td>R</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28119</th>\n",
       "      <td>R740Y</td>\n",
       "      <td>B</td>\n",
       "      <td>R740Y B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>-3.2</td>\n",
       "      <td>5.10</td>\n",
       "      <td>R</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28120 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      KatG mutations Chain Unique ID Distance from Site 1  d_volume  d_MW  \\\n",
       "0                M1A     A     M1A A    No centre of mass      74.3  60.1   \n",
       "1                M1C     A     M1C A    No centre of mass      54.4  28.1   \n",
       "2                M1D     A     M1D A    No centre of mass      51.8  16.1   \n",
       "3                M1E     A     M1E A    No centre of mass      24.5   2.1   \n",
       "4                M1F     A     M1F A    No centre of mass     -27.0 -16.0   \n",
       "...              ...   ...       ...                  ...       ...   ...   \n",
       "28115          R740S     B   R740S B            38.661813      84.4  69.1   \n",
       "28116          R740T     B   R740T B            38.661813      57.3  55.1   \n",
       "28117          R740V     B   R740V B            38.661813      33.4  57.1   \n",
       "28118          R740W     B   R740W B            38.661813     -54.4 -30.0   \n",
       "28119          R740Y     B   R740Y B            38.661813     -20.2  -7.0   \n",
       "\n",
       "       d_hydropathy  d_Pi Start residue End residue  \n",
       "0               0.1 -0.26             M           A  \n",
       "1              -0.6  0.67             M           C  \n",
       "2               5.4  2.97             M           D  \n",
       "3               5.4  2.52             M           E  \n",
       "4              -0.9  0.26             M           F  \n",
       "...             ...   ...           ...         ...  \n",
       "28115          -3.7  5.08             R           S  \n",
       "28116          -3.8  5.16             R           T  \n",
       "28117          -8.7  4.80             R           V  \n",
       "28118          -3.6  4.87             R           W  \n",
       "28119          -3.2  5.10             R           Y  \n",
       "\n",
       "[28120 rows x 10 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#adds start and end residues for each mutation to main df\n",
    "start_aa = []\n",
    "end_aa = []\n",
    "for i in KatG_mutations_df['KatG mutations']:\n",
    "    start = i[0]\n",
    "    end = i[-1]\n",
    "    start_aa.append(start)\n",
    "    end_aa.append(end)\n",
    "KatG_mutations_df['Start residue'] = start_aa\n",
    "KatG_mutations_df['End residue'] = end_aa\n",
    "KatG_mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "675fb286",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sets index before joining 2 dataframes\n",
    "codon_df.set_index('Start codon', inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f317a40f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start codon</th>\n",
       "      <th>End codon</th>\n",
       "      <th>Number of substitutions</th>\n",
       "      <th>Start AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>3</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAA</td>\n",
       "      <td>TTA</td>\n",
       "      <td>2</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAA</td>\n",
       "      <td>TTC</td>\n",
       "      <td>3</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAA</td>\n",
       "      <td>TTG</td>\n",
       "      <td>3</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAA</td>\n",
       "      <td>TAT</td>\n",
       "      <td>2</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4091</th>\n",
       "      <td>TTT</td>\n",
       "      <td>GCG</td>\n",
       "      <td>3</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4092</th>\n",
       "      <td>TTT</td>\n",
       "      <td>GGT</td>\n",
       "      <td>2</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4093</th>\n",
       "      <td>TTT</td>\n",
       "      <td>GGA</td>\n",
       "      <td>3</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4094</th>\n",
       "      <td>TTT</td>\n",
       "      <td>GGC</td>\n",
       "      <td>3</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4095</th>\n",
       "      <td>TTT</td>\n",
       "      <td>GGG</td>\n",
       "      <td>3</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4096 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Start codon End codon  Number of substitutions Start AA\n",
       "0            AAA       TTT                        3        K\n",
       "1            AAA       TTA                        2        K\n",
       "2            AAA       TTC                        3        K\n",
       "3            AAA       TTG                        3        K\n",
       "4            AAA       TAT                        2        K\n",
       "...          ...       ...                      ...      ...\n",
       "4091         TTT       GCG                        3        F\n",
       "4092         TTT       GGT                        2        F\n",
       "4093         TTT       GGA                        3        F\n",
       "4094         TTT       GGC                        3        F\n",
       "4095         TTT       GGG                        3        F\n",
       "\n",
       "[4096 rows x 4 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#creates dataframe which shows the starting amino acid for each codon mutation\n",
    "combined_codon_df = codon_df.join(aa_codons)\n",
    "combined_codon_df.reset_index(inplace = True)\n",
    "combined_codon_df.rename(columns = {'Amino Acid':'Start AA','index':'Start codon'}, inplace = True)\n",
    "combined_codon_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2dfe7ad2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>End codon</th>\n",
       "      <th>Start codon</th>\n",
       "      <th>Number of substitutions</th>\n",
       "      <th>Start AA</th>\n",
       "      <th>End AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAA</td>\n",
       "      <td>AAA</td>\n",
       "      <td>0</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAA</td>\n",
       "      <td>AAC</td>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAA</td>\n",
       "      <td>AAG</td>\n",
       "      <td>1</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAA</td>\n",
       "      <td>AAT</td>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAA</td>\n",
       "      <td>ACA</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4091</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TGT</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4092</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTA</td>\n",
       "      <td>1</td>\n",
       "      <td>L</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4093</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTC</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4094</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTG</td>\n",
       "      <td>1</td>\n",
       "      <td>L</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4095</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTT</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4096 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     End codon Start codon  Number of substitutions Start AA End AA\n",
       "0          AAA         AAA                        0        K      K\n",
       "1          AAA         AAC                        1        N      K\n",
       "2          AAA         AAG                        1        K      K\n",
       "3          AAA         AAT                        1        N      K\n",
       "4          AAA         ACA                        1        T      K\n",
       "...        ...         ...                      ...      ...    ...\n",
       "4091       TTT         TGT                        1        C      F\n",
       "4092       TTT         TTA                        1        L      F\n",
       "4093       TTT         TTC                        1        F      F\n",
       "4094       TTT         TTG                        1        L      F\n",
       "4095       TTT         TTT                        0        F      F\n",
       "\n",
       "[4096 rows x 5 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#adds end amino acid onto the dataframe\n",
    "combined_codon_df.set_index('End codon', inplace = True)\n",
    "combined_codon_df = combined_codon_df.join(aa_codons)\n",
    "combined_codon_df.reset_index(inplace = True)\n",
    "combined_codon_df.rename(columns = {'Amino Acid':'End AA','index':'End codon'}, inplace = True)\n",
    "combined_codon_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "7779437d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">Number of substitutions</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>mean</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Start AA</th>\n",
       "      <th>End AA</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">A</th>\n",
       "      <th>A</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2.750000</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2.750000</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Y</th>\n",
       "      <th>Stop</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.333333</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2.750000</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2.750000</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>441 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Number of substitutions                    \n",
       "                                    min max      mean count\n",
       "Start AA End AA                                            \n",
       "A        A                            0   1  0.750000    16\n",
       "         C                            2   3  2.750000     8\n",
       "         D                            1   2  1.750000     8\n",
       "         E                            1   2  1.750000     8\n",
       "         F                            2   3  2.750000     8\n",
       "...                                 ...  ..       ...   ...\n",
       "Y        Stop                         1   2  1.333333     6\n",
       "         T                            2   3  2.750000     8\n",
       "         V                            2   3  2.750000     8\n",
       "         W                            2   2  2.000000     2\n",
       "         Y                            0   1  0.500000     4\n",
       "\n",
       "[441 rows x 4 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#breaks down the number of substitutions based on the residue mutation\n",
    "combined_codon_df[['Start AA', 'End AA', 'Number of substitutions']].groupby(['Start AA','End AA']).agg(['min','max',numpy.mean,'count'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f09cb2a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>End codon</th>\n",
       "      <th>Start codon</th>\n",
       "      <th>Number of substitutions</th>\n",
       "      <th>Start AA</th>\n",
       "      <th>End AA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KK</th>\n",
       "      <td>AAA</td>\n",
       "      <td>AAA</td>\n",
       "      <td>0</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>AAA</td>\n",
       "      <td>AAC</td>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KK</th>\n",
       "      <td>AAA</td>\n",
       "      <td>AAG</td>\n",
       "      <td>1</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>AAA</td>\n",
       "      <td>AAT</td>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TK</th>\n",
       "      <td>AAA</td>\n",
       "      <td>ACA</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CF</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TGT</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTA</td>\n",
       "      <td>1</td>\n",
       "      <td>L</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FF</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTC</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTG</td>\n",
       "      <td>1</td>\n",
       "      <td>L</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FF</th>\n",
       "      <td>TTT</td>\n",
       "      <td>TTT</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4096 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         End codon Start codon  Number of substitutions Start AA End AA\n",
       "Mutation                                                               \n",
       "KK             AAA         AAA                        0        K      K\n",
       "NK             AAA         AAC                        1        N      K\n",
       "KK             AAA         AAG                        1        K      K\n",
       "NK             AAA         AAT                        1        N      K\n",
       "TK             AAA         ACA                        1        T      K\n",
       "...            ...         ...                      ...      ...    ...\n",
       "CF             TTT         TGT                        1        C      F\n",
       "LF             TTT         TTA                        1        L      F\n",
       "FF             TTT         TTC                        1        F      F\n",
       "LF             TTT         TTG                        1        L      F\n",
       "FF             TTT         TTT                        0        F      F\n",
       "\n",
       "[4096 rows x 5 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#adds start and end amino acid together in 1 column to get the mutation \n",
    "mutation_1 = []\n",
    "row_no_1 = -1\n",
    "for s_aa in combined_codon_df['Start AA']:\n",
    "    row_no_1 = row_no_1 + 1 \n",
    "    e_aa = combined_codon_df.loc[row_no_1,'End AA']\n",
    "    d_aa = str(s_aa + e_aa)\n",
    "    mutation_1.append(d_aa)\n",
    "combined_codon_df['Mutation'] = mutation_1\n",
    "combined_codon_df.set_index('Mutation', inplace = True)\n",
    "combined_codon_df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "c2992dcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KatG mutations</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Unique ID</th>\n",
       "      <th>Distance from Site 1</th>\n",
       "      <th>d_volume</th>\n",
       "      <th>d_MW</th>\n",
       "      <th>d_hydropathy</th>\n",
       "      <th>d_Pi</th>\n",
       "      <th>Start residue</th>\n",
       "      <th>End residue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MA</th>\n",
       "      <td>M1A</td>\n",
       "      <td>A</td>\n",
       "      <td>M1A A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>74.3</td>\n",
       "      <td>60.1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>-0.26</td>\n",
       "      <td>M</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MC</th>\n",
       "      <td>M1C</td>\n",
       "      <td>A</td>\n",
       "      <td>M1C A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>54.4</td>\n",
       "      <td>28.1</td>\n",
       "      <td>-0.6</td>\n",
       "      <td>0.67</td>\n",
       "      <td>M</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MD</th>\n",
       "      <td>M1D</td>\n",
       "      <td>A</td>\n",
       "      <td>M1D A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>51.8</td>\n",
       "      <td>16.1</td>\n",
       "      <td>5.4</td>\n",
       "      <td>2.97</td>\n",
       "      <td>M</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ME</th>\n",
       "      <td>M1E</td>\n",
       "      <td>A</td>\n",
       "      <td>M1E A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>24.5</td>\n",
       "      <td>2.1</td>\n",
       "      <td>5.4</td>\n",
       "      <td>2.52</td>\n",
       "      <td>M</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MF</th>\n",
       "      <td>M1F</td>\n",
       "      <td>A</td>\n",
       "      <td>M1F A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>-27.0</td>\n",
       "      <td>-16.0</td>\n",
       "      <td>-0.9</td>\n",
       "      <td>0.26</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS</th>\n",
       "      <td>R740S</td>\n",
       "      <td>B</td>\n",
       "      <td>R740S B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>84.4</td>\n",
       "      <td>69.1</td>\n",
       "      <td>-3.7</td>\n",
       "      <td>5.08</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RT</th>\n",
       "      <td>R740T</td>\n",
       "      <td>B</td>\n",
       "      <td>R740T B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>57.3</td>\n",
       "      <td>55.1</td>\n",
       "      <td>-3.8</td>\n",
       "      <td>5.16</td>\n",
       "      <td>R</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RV</th>\n",
       "      <td>R740V</td>\n",
       "      <td>B</td>\n",
       "      <td>R740V B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>33.4</td>\n",
       "      <td>57.1</td>\n",
       "      <td>-8.7</td>\n",
       "      <td>4.80</td>\n",
       "      <td>R</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RW</th>\n",
       "      <td>R740W</td>\n",
       "      <td>B</td>\n",
       "      <td>R740W B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>-54.4</td>\n",
       "      <td>-30.0</td>\n",
       "      <td>-3.6</td>\n",
       "      <td>4.87</td>\n",
       "      <td>R</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RY</th>\n",
       "      <td>R740Y</td>\n",
       "      <td>B</td>\n",
       "      <td>R740Y B</td>\n",
       "      <td>38.661813</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>-3.2</td>\n",
       "      <td>5.10</td>\n",
       "      <td>R</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28120 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         KatG mutations Chain Unique ID Distance from Site 1  d_volume  d_MW  \\\n",
       "Mutation                                                                       \n",
       "MA                  M1A     A     M1A A    No centre of mass      74.3  60.1   \n",
       "MC                  M1C     A     M1C A    No centre of mass      54.4  28.1   \n",
       "MD                  M1D     A     M1D A    No centre of mass      51.8  16.1   \n",
       "ME                  M1E     A     M1E A    No centre of mass      24.5   2.1   \n",
       "MF                  M1F     A     M1F A    No centre of mass     -27.0 -16.0   \n",
       "...                 ...   ...       ...                  ...       ...   ...   \n",
       "RS                R740S     B   R740S B            38.661813      84.4  69.1   \n",
       "RT                R740T     B   R740T B            38.661813      57.3  55.1   \n",
       "RV                R740V     B   R740V B            38.661813      33.4  57.1   \n",
       "RW                R740W     B   R740W B            38.661813     -54.4 -30.0   \n",
       "RY                R740Y     B   R740Y B            38.661813     -20.2  -7.0   \n",
       "\n",
       "          d_hydropathy  d_Pi Start residue End residue  \n",
       "Mutation                                                \n",
       "MA                 0.1 -0.26             M           A  \n",
       "MC                -0.6  0.67             M           C  \n",
       "MD                 5.4  2.97             M           D  \n",
       "ME                 5.4  2.52             M           E  \n",
       "MF                -0.9  0.26             M           F  \n",
       "...                ...   ...           ...         ...  \n",
       "RS                -3.7  5.08             R           S  \n",
       "RT                -3.8  5.16             R           T  \n",
       "RV                -8.7  4.80             R           V  \n",
       "RW                -3.6  4.87             R           W  \n",
       "RY                -3.2  5.10             R           Y  \n",
       "\n",
       "[28120 rows x 10 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#adds start and end amino acid together to get the mutation\n",
    "KatG_mutations_df\n",
    "mutation_2 = []\n",
    "row_no_2 = -1 \n",
    "for s_aa in KatG_mutations_df['Start residue']:\n",
    "    row_no_2 = row_no_2 + 1\n",
    "    e_aa = KatG_mutations_df.loc[row_no_2,'End residue']\n",
    "    d_aa = str(s_aa + e_aa)\n",
    "    mutation_2.append(d_aa)\n",
    "KatG_mutations_df['Mutation'] = mutation_2\n",
    "KatG_mutations_df.set_index('Mutation', inplace = True)\n",
    "KatG_mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6a2a59e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"3\" halign=\"left\">Number of substitutions</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AD</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AE</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YStop</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YT</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YV</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YW</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YY</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>441 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Number of substitutions          \n",
       "                             min max count\n",
       "Mutation                                  \n",
       "AA                             0   1    16\n",
       "AC                             2   3     8\n",
       "AD                             1   2     8\n",
       "AE                             1   2     8\n",
       "AF                             2   3     8\n",
       "...                          ...  ..   ...\n",
       "YStop                          1   2     6\n",
       "YT                             2   3     8\n",
       "YV                             2   3     8\n",
       "YW                             2   2     2\n",
       "YY                             0   1     4\n",
       "\n",
       "[441 rows x 3 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#creates df to show how the number of substitutions varies for each mutation\n",
    "combined_codon_df.reset_index(inplace = True)\n",
    "subs_df = combined_codon_df[['Mutation', 'Number of substitutions']].groupby(['Mutation']).agg(['min','max','count'])\n",
    "subs_df.reset_index(inplace = True)\n",
    "subs_df.set_index('Mutation', inplace = True)\n",
    "subs_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "242a5e96",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\user\\anaconda3\\lib\\site-packages\\pandas\\core\\frame.py:9130: FutureWarning: merging between different levels is deprecated and will be removed in a future version. (1 levels on the left,2 on the right)\n",
      "  return merge(\n"
     ]
    }
   ],
   "source": [
    "#adds number of substitutions for each possible mutation in the main df\n",
    "KatG_mutations_df = KatG_mutations_df.join(subs_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b6c2fe16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KatG mutations</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Unique ID</th>\n",
       "      <th>Distance from Site 1</th>\n",
       "      <th>d_volume</th>\n",
       "      <th>d_MW</th>\n",
       "      <th>d_hydropathy</th>\n",
       "      <th>d_Pi</th>\n",
       "      <th>Start residue</th>\n",
       "      <th>End residue</th>\n",
       "      <th>(Number of substitutions, min)</th>\n",
       "      <th>(Number of substitutions, max)</th>\n",
       "      <th>(Number of substitutions, count)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AC</th>\n",
       "      <td>A15C</td>\n",
       "      <td>A</td>\n",
       "      <td>A15C A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC</th>\n",
       "      <td>A16C</td>\n",
       "      <td>A</td>\n",
       "      <td>A16C A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC</th>\n",
       "      <td>A53C</td>\n",
       "      <td>A</td>\n",
       "      <td>A53C A</td>\n",
       "      <td>35.218287</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC</th>\n",
       "      <td>A55C</td>\n",
       "      <td>A</td>\n",
       "      <td>A55C A</td>\n",
       "      <td>31.791704</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC</th>\n",
       "      <td>A60C</td>\n",
       "      <td>A</td>\n",
       "      <td>A60C A</td>\n",
       "      <td>38.926044</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YW</th>\n",
       "      <td>Y597W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y597W B</td>\n",
       "      <td>40.613433</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YW</th>\n",
       "      <td>Y608W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y608W B</td>\n",
       "      <td>33.528173</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YW</th>\n",
       "      <td>Y638W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y638W B</td>\n",
       "      <td>46.007273</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YW</th>\n",
       "      <td>Y678W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y678W B</td>\n",
       "      <td>48.498844</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YW</th>\n",
       "      <td>Y711W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y711W B</td>\n",
       "      <td>50.160048</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28120 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         KatG mutations Chain Unique ID Distance from Site 1  d_volume  d_MW  \\\n",
       "Mutation                                                                       \n",
       "AC                 A15C     A    A15C A    No centre of mass     -19.9 -32.0   \n",
       "AC                 A16C     A    A16C A    No centre of mass     -19.9 -32.0   \n",
       "AC                 A53C     A    A53C A            35.218287     -19.9 -32.0   \n",
       "AC                 A55C     A    A55C A            31.791704     -19.9 -32.0   \n",
       "AC                 A60C     A    A60C A            38.926044     -19.9 -32.0   \n",
       "...                 ...   ...       ...                  ...       ...   ...   \n",
       "YW                Y597W     B   Y597W B            40.613433     -34.2 -23.0   \n",
       "YW                Y608W     B   Y608W B            33.528173     -34.2 -23.0   \n",
       "YW                Y638W     B   Y638W B            46.007273     -34.2 -23.0   \n",
       "YW                Y678W     B   Y678W B            48.498844     -34.2 -23.0   \n",
       "YW                Y711W     B   Y711W B            50.160048     -34.2 -23.0   \n",
       "\n",
       "          d_hydropathy  d_Pi Start residue End residue  \\\n",
       "Mutation                                                 \n",
       "AC                -0.7  0.93             A           C   \n",
       "AC                -0.7  0.93             A           C   \n",
       "AC                -0.7  0.93             A           C   \n",
       "AC                -0.7  0.93             A           C   \n",
       "AC                -0.7  0.93             A           C   \n",
       "...                ...   ...           ...         ...   \n",
       "YW                -0.4 -0.23             Y           W   \n",
       "YW                -0.4 -0.23             Y           W   \n",
       "YW                -0.4 -0.23             Y           W   \n",
       "YW                -0.4 -0.23             Y           W   \n",
       "YW                -0.4 -0.23             Y           W   \n",
       "\n",
       "          (Number of substitutions, min)  (Number of substitutions, max)  \\\n",
       "Mutation                                                                   \n",
       "AC                                     2                               3   \n",
       "AC                                     2                               3   \n",
       "AC                                     2                               3   \n",
       "AC                                     2                               3   \n",
       "AC                                     2                               3   \n",
       "...                                  ...                             ...   \n",
       "YW                                     2                               2   \n",
       "YW                                     2                               2   \n",
       "YW                                     2                               2   \n",
       "YW                                     2                               2   \n",
       "YW                                     2                               2   \n",
       "\n",
       "          (Number of substitutions, count)  \n",
       "Mutation                                    \n",
       "AC                                       8  \n",
       "AC                                       8  \n",
       "AC                                       8  \n",
       "AC                                       8  \n",
       "AC                                       8  \n",
       "...                                    ...  \n",
       "YW                                       2  \n",
       "YW                                       2  \n",
       "YW                                       2  \n",
       "YW                                       2  \n",
       "YW                                       2  \n",
       "\n",
       "[28120 rows x 13 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KatG_mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "658db9bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation</th>\n",
       "      <th>KatG mutations</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Unique ID</th>\n",
       "      <th>Distance from Site 1</th>\n",
       "      <th>d_volume</th>\n",
       "      <th>d_MW</th>\n",
       "      <th>d_hydropathy</th>\n",
       "      <th>d_Pi</th>\n",
       "      <th>Start residue</th>\n",
       "      <th>End residue</th>\n",
       "      <th>(Number of substitutions, min)</th>\n",
       "      <th>(Number of substitutions, max)</th>\n",
       "      <th>(Number of substitutions, count)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AC</td>\n",
       "      <td>A15C</td>\n",
       "      <td>A</td>\n",
       "      <td>A15C A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC</td>\n",
       "      <td>A16C</td>\n",
       "      <td>A</td>\n",
       "      <td>A16C A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AC</td>\n",
       "      <td>A53C</td>\n",
       "      <td>A</td>\n",
       "      <td>A53C A</td>\n",
       "      <td>35.218287</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AC</td>\n",
       "      <td>A55C</td>\n",
       "      <td>A</td>\n",
       "      <td>A55C A</td>\n",
       "      <td>31.791704</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AC</td>\n",
       "      <td>A60C</td>\n",
       "      <td>A</td>\n",
       "      <td>A60C A</td>\n",
       "      <td>38.926044</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28115</th>\n",
       "      <td>YW</td>\n",
       "      <td>Y597W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y597W B</td>\n",
       "      <td>40.613433</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28116</th>\n",
       "      <td>YW</td>\n",
       "      <td>Y608W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y608W B</td>\n",
       "      <td>33.528173</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28117</th>\n",
       "      <td>YW</td>\n",
       "      <td>Y638W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y638W B</td>\n",
       "      <td>46.007273</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28118</th>\n",
       "      <td>YW</td>\n",
       "      <td>Y678W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y678W B</td>\n",
       "      <td>48.498844</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28119</th>\n",
       "      <td>YW</td>\n",
       "      <td>Y711W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y711W B</td>\n",
       "      <td>50.160048</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28120 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Mutation KatG mutations Chain Unique ID Distance from Site 1  d_volume  \\\n",
       "0           AC           A15C     A    A15C A    No centre of mass     -19.9   \n",
       "1           AC           A16C     A    A16C A    No centre of mass     -19.9   \n",
       "2           AC           A53C     A    A53C A            35.218287     -19.9   \n",
       "3           AC           A55C     A    A55C A            31.791704     -19.9   \n",
       "4           AC           A60C     A    A60C A            38.926044     -19.9   \n",
       "...        ...            ...   ...       ...                  ...       ...   \n",
       "28115       YW          Y597W     B   Y597W B            40.613433     -34.2   \n",
       "28116       YW          Y608W     B   Y608W B            33.528173     -34.2   \n",
       "28117       YW          Y638W     B   Y638W B            46.007273     -34.2   \n",
       "28118       YW          Y678W     B   Y678W B            48.498844     -34.2   \n",
       "28119       YW          Y711W     B   Y711W B            50.160048     -34.2   \n",
       "\n",
       "       d_MW  d_hydropathy  d_Pi Start residue End residue  \\\n",
       "0     -32.0          -0.7  0.93             A           C   \n",
       "1     -32.0          -0.7  0.93             A           C   \n",
       "2     -32.0          -0.7  0.93             A           C   \n",
       "3     -32.0          -0.7  0.93             A           C   \n",
       "4     -32.0          -0.7  0.93             A           C   \n",
       "...     ...           ...   ...           ...         ...   \n",
       "28115 -23.0          -0.4 -0.23             Y           W   \n",
       "28116 -23.0          -0.4 -0.23             Y           W   \n",
       "28117 -23.0          -0.4 -0.23             Y           W   \n",
       "28118 -23.0          -0.4 -0.23             Y           W   \n",
       "28119 -23.0          -0.4 -0.23             Y           W   \n",
       "\n",
       "       (Number of substitutions, min)  (Number of substitutions, max)  \\\n",
       "0                                   2                               3   \n",
       "1                                   2                               3   \n",
       "2                                   2                               3   \n",
       "3                                   2                               3   \n",
       "4                                   2                               3   \n",
       "...                               ...                             ...   \n",
       "28115                               2                               2   \n",
       "28116                               2                               2   \n",
       "28117                               2                               2   \n",
       "28118                               2                               2   \n",
       "28119                               2                               2   \n",
       "\n",
       "       (Number of substitutions, count)  \n",
       "0                                     8  \n",
       "1                                     8  \n",
       "2                                     8  \n",
       "3                                     8  \n",
       "4                                     8  \n",
       "...                                 ...  \n",
       "28115                                 2  \n",
       "28116                                 2  \n",
       "28117                                 2  \n",
       "28118                                 2  \n",
       "28119                                 2  \n",
       "\n",
       "[28120 rows x 14 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#resets main index on main df\n",
    "KatG_mutations_df.reset_index(inplace=True)\n",
    "KatG_mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b8409177",
   "metadata": {},
   "outputs": [],
   "source": [
    "#freesasa code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b6f994f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\user\\anaconda3\\lib\\site-packages\\MDAnalysis\\coordinates\\PDB.py:429: UserWarning: 1 A^3 CRYST1 record, this is usually a placeholder. Unit cell dimensions will be set to None.\n",
      "  warnings.warn(\"1 A^3 CRYST1 record,\"\n"
     ]
    }
   ],
   "source": [
    "#calculates SASA for each atom in KatG using edited PDB with only CA altLoc B removed\n",
    "KatG_edit = mda.Universe(\"7ag8_R418_edit.pdb\")\n",
    "structure = freesasa.Structure(\"7ag8_R418_edit.pdb\")\n",
    "result = freesasa.calc(structure)\n",
    "\n",
    "ASA_dict_atom = {}\n",
    "for i in range(0, structure.nAtoms()):\n",
    "    ASA_dict_atom[i] = result.atomArea(i)\n",
    "    \n",
    "SASA_df = pd.DataFrame.from_dict(ASA_dict_atom, orient='index', columns = ['atom_sasa'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "0e5b832d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dict for each residues code\n",
    "res_codes = {'ALA':'A', 'ARG':'R', 'ASN':'N', 'ASP':'D', 'CYS':'C', \n",
    "             'GLU':'E', 'GLN':'Q', 'GLY':'G', 'HIS':'H', 'ILE':'I', \n",
    "             'LEU':'L', 'LYS':'K', 'MET':'M', 'PHE':'F', 'PRO':'P', \n",
    "             'SER':'S', 'THR':'T', 'TRP':'W', 'TYR':'Y', 'VAL':'V'}\n",
    "\n",
    "#define max total sasa for each residue (also useful to define aa)\n",
    "maxasa_dict = {'ALA':129, 'ARG':274, 'ASN':195, 'ASP':193, 'CYS':167, 'GLU':223,\n",
    "                'GLN':225, 'GLY':104, 'HIS':224, 'ILE':197, 'LEU':201, 'LYS':236,\n",
    "                'MET':224, 'PHE':240, 'PRO':159, 'SER':155, 'THR':172, 'TRP':285,\n",
    "                'TYR':263, 'VAL':174}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "25676004",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>atom_sasa</th>\n",
       "      <th>atom</th>\n",
       "      <th>residue</th>\n",
       "      <th>residue_number</th>\n",
       "      <th>chain</th>\n",
       "      <th>combined_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26.778193</td>\n",
       "      <td>N</td>\n",
       "      <td>GLY</td>\n",
       "      <td>24</td>\n",
       "      <td>A</td>\n",
       "      <td>A_0024_GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>41.526774</td>\n",
       "      <td>CA</td>\n",
       "      <td>GLY</td>\n",
       "      <td>24</td>\n",
       "      <td>A</td>\n",
       "      <td>A_0024_GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.935937</td>\n",
       "      <td>C</td>\n",
       "      <td>GLY</td>\n",
       "      <td>24</td>\n",
       "      <td>A</td>\n",
       "      <td>A_0024_GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30.605271</td>\n",
       "      <td>O</td>\n",
       "      <td>GLY</td>\n",
       "      <td>24</td>\n",
       "      <td>A</td>\n",
       "      <td>A_0024_GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>N</td>\n",
       "      <td>HIS</td>\n",
       "      <td>25</td>\n",
       "      <td>A</td>\n",
       "      <td>A_0025_HIS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10815</th>\n",
       "      <td>6.324353</td>\n",
       "      <td>NE</td>\n",
       "      <td>ARG</td>\n",
       "      <td>740</td>\n",
       "      <td>B</td>\n",
       "      <td>B_0740_ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10816</th>\n",
       "      <td>1.726302</td>\n",
       "      <td>CZ</td>\n",
       "      <td>ARG</td>\n",
       "      <td>740</td>\n",
       "      <td>B</td>\n",
       "      <td>B_0740_ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10817</th>\n",
       "      <td>2.320039</td>\n",
       "      <td>NH1</td>\n",
       "      <td>ARG</td>\n",
       "      <td>740</td>\n",
       "      <td>B</td>\n",
       "      <td>B_0740_ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10818</th>\n",
       "      <td>20.751627</td>\n",
       "      <td>NH2</td>\n",
       "      <td>ARG</td>\n",
       "      <td>740</td>\n",
       "      <td>B</td>\n",
       "      <td>B_0740_ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10819</th>\n",
       "      <td>35.984040</td>\n",
       "      <td>OXT</td>\n",
       "      <td>ARG</td>\n",
       "      <td>740</td>\n",
       "      <td>B</td>\n",
       "      <td>B_0740_ARG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10820 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       atom_sasa  atom residue residue_number chain combined_ID\n",
       "0      26.778193   N       GLY            24      A  A_0024_GLY\n",
       "1      41.526774   CA      GLY            24      A  A_0024_GLY\n",
       "2       1.935937   C       GLY            24      A  A_0024_GLY\n",
       "3      30.605271   O       GLY            24      A  A_0024_GLY\n",
       "4       0.000000   N       HIS            25      A  A_0025_HIS\n",
       "...          ...   ...     ...            ...   ...         ...\n",
       "10815   6.324353   NE      ARG           740      B  B_0740_ARG\n",
       "10816   1.726302   CZ      ARG           740      B  B_0740_ARG\n",
       "10817   2.320039   NH1     ARG           740      B  B_0740_ARG\n",
       "10818  20.751627   NH2     ARG           740      B  B_0740_ARG\n",
       "10819  35.984040   OXT     ARG           740      B  B_0740_ARG\n",
       "\n",
       "[10820 rows x 6 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "    #creates df with each atoms SASA and a unique combined ID \n",
    "    residue_list, atom_list, residue_number, chain_label, combined_ID = [], [], [], [], []\n",
    "    for i in range(0, structure.nAtoms()):\n",
    "        residue_list.append(structure.residueName(i))\n",
    "        atom_list.append(structure.atomName(i))\n",
    "        residue_number.append(structure.residueNumber(i))\n",
    "        chain_label.append(structure.chainLabel(i))\n",
    "        j = 0\n",
    "        if int(structure.residueNumber(i)) < 10:\n",
    "            j = '000' + str(structure.residueNumber(i)).strip()\n",
    "        elif int(structure.residueNumber(i)) < 100:\n",
    "            j = '00' + str(structure.residueNumber(i)).strip()\n",
    "        elif int(structure.residueNumber(i)) < 1000:\n",
    "            j = '0' + str(structure.residueNumber(i)).strip()\n",
    "        else:\n",
    "            j = structure.residueNumber(i).strip()\n",
    "        combined_ID.append(structure.chainLabel(i) + '_' + str(j) + '_' + structure.residueName(i))\n",
    "\n",
    "    SASA_df['atom'] = atom_list\n",
    "    SASA_df['residue'] = residue_list\n",
    "    SASA_df['residue_number'] = residue_number\n",
    "    SASA_df['chain'] = chain_label\n",
    "    SASA_df['combined_ID'] = combined_ID\n",
    "    \n",
    "    SASA_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "f9ea5a1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>combined_ID</th>\n",
       "      <th>SASA_residue</th>\n",
       "      <th>residue_ID</th>\n",
       "      <th>CA_ID</th>\n",
       "      <th>maxASA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_0024_GLY</td>\n",
       "      <td>100.846174</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_0025_HIS</td>\n",
       "      <td>127.637993</td>\n",
       "      <td>25</td>\n",
       "      <td>5</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_0026_MET</td>\n",
       "      <td>18.262181</td>\n",
       "      <td>26</td>\n",
       "      <td>15</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_0027_LYS</td>\n",
       "      <td>61.530393</td>\n",
       "      <td>27</td>\n",
       "      <td>23</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_0028_TYR</td>\n",
       "      <td>39.727088</td>\n",
       "      <td>28</td>\n",
       "      <td>32</td>\n",
       "      <td>263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1412</th>\n",
       "      <td>B_0736_ARG</td>\n",
       "      <td>0.125640</td>\n",
       "      <td>736</td>\n",
       "      <td>10772</td>\n",
       "      <td>274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1413</th>\n",
       "      <td>B_0737_PHE</td>\n",
       "      <td>34.730453</td>\n",
       "      <td>737</td>\n",
       "      <td>10783</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1414</th>\n",
       "      <td>B_0738_ASP</td>\n",
       "      <td>79.101138</td>\n",
       "      <td>738</td>\n",
       "      <td>10794</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1415</th>\n",
       "      <td>B_0739_VAL</td>\n",
       "      <td>81.799088</td>\n",
       "      <td>739</td>\n",
       "      <td>10802</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1416</th>\n",
       "      <td>B_0740_ARG</td>\n",
       "      <td>161.555499</td>\n",
       "      <td>740</td>\n",
       "      <td>10809</td>\n",
       "      <td>274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1417 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     combined_ID  SASA_residue residue_ID  CA_ID  maxASA\n",
       "0     A_0024_GLY    100.846174        24       1     104\n",
       "1     A_0025_HIS    127.637993        25       5     224\n",
       "2     A_0026_MET     18.262181        26      15     224\n",
       "3     A_0027_LYS     61.530393        27      23     236\n",
       "4     A_0028_TYR     39.727088        28      32     263\n",
       "...          ...           ...        ...    ...     ...\n",
       "1412  B_0736_ARG      0.125640       736   10772     274\n",
       "1413  B_0737_PHE     34.730453       737   10783     240\n",
       "1414  B_0738_ASP     79.101138       738   10794     193\n",
       "1415  B_0739_VAL     81.799088       739   10802     174\n",
       "1416  B_0740_ARG    161.555499       740   10809     274\n",
       "\n",
       "[1417 rows x 5 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "    #creates dataframe with the sum of the SASA of all atoms in each residue\n",
    "    residue_list, residue_ID_list, CA_atom_ID_list = [], [], []\n",
    "    Dict = {}\n",
    "    SASA_sum = 0\n",
    "\n",
    "    for i in SASA_df.index:\n",
    "        try:\n",
    "            if SASA_df['combined_ID'][i] == SASA_df['combined_ID'][i+1]:\n",
    "                #atoms in same residue, so add current sasa and keep iterating\n",
    "                SASA_sum += SASA_df['atom_sasa'][i]\n",
    "            else:\n",
    "                #atoms not in same residue, so add sasa and reset SASA_sum\n",
    "                SASA_sum += SASA_df['atom_sasa'][i]\n",
    "                Dict[SASA_df['combined_ID'][i]] = SASA_sum\n",
    "                SASA_sum = 0\n",
    "                residue_list.append(SASA_df['residue'][i])\n",
    "                residue_ID_list.append(SASA_df['residue_number'][i])\n",
    "        except KeyError:\n",
    "            #for last index of df, so add sasa and reset SASA_sum\n",
    "            SASA_sum += SASA_df['atom_sasa'][i]\n",
    "            Dict[SASA_df['combined_ID'][i]] = SASA_sum\n",
    "            SASA_sum = 0\n",
    "            residue_list.append(SASA_df['residue'][i])\n",
    "            residue_ID_list.append(SASA_df['residue_number'][i])\n",
    "        if SASA_df['atom'][i].strip() == 'CA':\n",
    "            #list indexes of CA \n",
    "            CA_atom_ID_list.append(SASA_df.index[i])\n",
    "    \n",
    "    sasa_sum_df = pd.DataFrame.from_dict(Dict, orient='index', \n",
    "                columns=['SASA_residue']).reset_index(0).rename(columns={'index':'combined_ID'})\n",
    "    sasa_sum_df['residue_ID'] = residue_ID_list\n",
    "    sasa_sum_df['CA_ID'] = CA_atom_ID_list\n",
    "    \n",
    "    #insert predefined max sasa for each residue\n",
    "    maxasa_list = []\n",
    "    for i in sasa_sum_df['combined_ID']:\n",
    "        for key, value in maxasa_dict.items():\n",
    "            if i[7:] == key:\n",
    "                maxasa_list.append(value)\n",
    "    sasa_sum_df['maxASA'] = maxasa_list\n",
    "    \n",
    "    sasa_sum_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c502d962",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>combined_ID</th>\n",
       "      <th>SASA_residue</th>\n",
       "      <th>residue_ID</th>\n",
       "      <th>CA_ID</th>\n",
       "      <th>maxASA</th>\n",
       "      <th>RSA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A_0024_GLY</td>\n",
       "      <td>100.846174</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>104</td>\n",
       "      <td>0.969675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A_0025_HIS</td>\n",
       "      <td>127.637993</td>\n",
       "      <td>25</td>\n",
       "      <td>5</td>\n",
       "      <td>224</td>\n",
       "      <td>0.569812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A_0027_LYS</td>\n",
       "      <td>61.530393</td>\n",
       "      <td>27</td>\n",
       "      <td>23</td>\n",
       "      <td>236</td>\n",
       "      <td>0.260722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>A_0037_ASP</td>\n",
       "      <td>53.610179</td>\n",
       "      <td>37</td>\n",
       "      <td>96</td>\n",
       "      <td>193</td>\n",
       "      <td>0.277773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>A_0040_PRO</td>\n",
       "      <td>77.247684</td>\n",
       "      <td>40</td>\n",
       "      <td>132</td>\n",
       "      <td>159</td>\n",
       "      <td>0.485834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>1402</td>\n",
       "      <td>B_0726_ALA</td>\n",
       "      <td>63.733677</td>\n",
       "      <td>726</td>\n",
       "      <td>10692</td>\n",
       "      <td>129</td>\n",
       "      <td>0.494060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>1406</td>\n",
       "      <td>B_0730_LYS</td>\n",
       "      <td>65.592781</td>\n",
       "      <td>730</td>\n",
       "      <td>10724</td>\n",
       "      <td>236</td>\n",
       "      <td>0.277936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>1414</td>\n",
       "      <td>B_0738_ASP</td>\n",
       "      <td>79.101138</td>\n",
       "      <td>738</td>\n",
       "      <td>10794</td>\n",
       "      <td>193</td>\n",
       "      <td>0.409850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>1415</td>\n",
       "      <td>B_0739_VAL</td>\n",
       "      <td>81.799088</td>\n",
       "      <td>739</td>\n",
       "      <td>10802</td>\n",
       "      <td>174</td>\n",
       "      <td>0.470110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>1416</td>\n",
       "      <td>B_0740_ARG</td>\n",
       "      <td>161.555499</td>\n",
       "      <td>740</td>\n",
       "      <td>10809</td>\n",
       "      <td>274</td>\n",
       "      <td>0.589619</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>423 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index combined_ID  SASA_residue residue_ID  CA_ID  maxASA       RSA\n",
       "0        0  A_0024_GLY    100.846174        24       1     104  0.969675\n",
       "1        1  A_0025_HIS    127.637993        25       5     224  0.569812\n",
       "2        3  A_0027_LYS     61.530393        27      23     236  0.260722\n",
       "3       13  A_0037_ASP     53.610179        37      96     193  0.277773\n",
       "4       16  A_0040_PRO     77.247684        40     132     159  0.485834\n",
       "..     ...         ...           ...        ...    ...     ...       ...\n",
       "418   1402  B_0726_ALA     63.733677       726   10692     129  0.494060\n",
       "419   1406  B_0730_LYS     65.592781       730   10724     236  0.277936\n",
       "420   1414  B_0738_ASP     79.101138       738   10794     193  0.409850\n",
       "421   1415  B_0739_VAL     81.799088       739   10802     174  0.470110\n",
       "422   1416  B_0740_ARG    161.555499       740   10809     274  0.589619\n",
       "\n",
       "[423 rows x 7 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "    #calculate and insert relative SASA for each residue\n",
    "    sasa_sum_df['RSA'] = sasa_sum_df['SASA_residue']/sasa_sum_df['maxASA']\n",
    "    #filter for residues with SASA > 0.25 as surface residues -- based on KatG surface area residues and RNAP calc\n",
    "    surface_residues_df = sasa_sum_df[sasa_sum_df['RSA'] >= 0.26].reset_index(0)\n",
    "    surface_residues_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "59c20621",
   "metadata": {},
   "outputs": [],
   "source": [
    "    #identifies how close each non surface alpha carbon is from a surface alpha carbon, denoted as depth\n",
    "    CA = KatG_edit.select_atoms('name CA')\n",
    "    sa_CA = AtomGroup(surface_residues_df['CA_ID'], KatG_edit)\n",
    "    distances = distance_array(CA.positions, sa_CA.positions)\n",
    "    min_distances = []\n",
    "    for i in distances:\n",
    "        min_distances.append(min(i))\n",
    "distance_df = pd.DataFrame(min_distances, columns=['Depth'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "88a0c957",
   "metadata": {},
   "outputs": [],
   "source": [
    "    #insert chain and numbering\n",
    "    chain_list, pdb_residue_list, resname_list = [], [], []\n",
    "    for i in CA:\n",
    "        chain_list.append(str(i.segment)[9])\n",
    "        pdb_residue_list.append(i.resid)\n",
    "        resname_list.append(i.resname)\n",
    "    #create and fill depth dataframe\n",
    "    distance_df['pdb_residue'] = pdb_residue_list\n",
    "    distance_df['pdb_chain'] = chain_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "69529273",
   "metadata": {},
   "outputs": [],
   "source": [
    "    #converts resnames to rescode IDs (ALA --> A e.g)\n",
    "    rescode_list = []\n",
    "    for i in resname_list:\n",
    "        for k, v in res_codes.items():\n",
    "            if i == k:\n",
    "                rescode_list.append(v)\n",
    "    distance_df['rescode'] = rescode_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "4332589e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adds unique ID to dpeth dataframe\n",
    "distance_df_unique_ID = []\n",
    "distance_df_row_no = -1\n",
    "for i in distance_df['Depth']:\n",
    "    distance_df_row_no += 1\n",
    "    Res = distance_df.loc[distance_df_row_no,'rescode']\n",
    "    Num = distance_df.loc[distance_df_row_no,'pdb_residue']\n",
    "    Chain = distance_df.loc[distance_df_row_no,'pdb_chain']\n",
    "    Unique = str(Res+str(Num)+'_'+Chain)\n",
    "    distance_df_unique_ID.append(Unique)\n",
    "distance_df['Residue Unique ID'] = distance_df_unique_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6fa9a583",
   "metadata": {},
   "outputs": [],
   "source": [
    "KatG_mutations_df.rename(columns={'Unique ID':'Mutation Unique ID'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "a75a3acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adds comparable unique ID to main df\n",
    "Residue_unique_ID = []\n",
    "for i in KatG_mutations_df['Mutation Unique ID']:\n",
    "    Res = str(i)[:-3]\n",
    "    Chain = str(i)[-1]\n",
    "    Unique = str(Res+'_'+Chain)\n",
    "    Residue_unique_ID.append(Unique)\n",
    "    \n",
    "KatG_mutations_df['Residue Unique ID'] = Residue_unique_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "b9ecb4d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#removes characteristic columns in depth df so only unique ID and depth are left\n",
    "distance_merge = distance_df.copy()\n",
    "del distance_merge['pdb_residue'],distance_merge['pdb_chain'],distance_merge['rescode']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b275dd37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation</th>\n",
       "      <th>KatG mutations</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Mutation Unique ID</th>\n",
       "      <th>Distance from Site 1</th>\n",
       "      <th>d_volume</th>\n",
       "      <th>d_MW</th>\n",
       "      <th>d_hydropathy</th>\n",
       "      <th>d_Pi</th>\n",
       "      <th>Start residue</th>\n",
       "      <th>End residue</th>\n",
       "      <th>(Number of substitutions, min)</th>\n",
       "      <th>(Number of substitutions, max)</th>\n",
       "      <th>(Number of substitutions, count)</th>\n",
       "      <th>Residue Unique ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AC</td>\n",
       "      <td>A15C</td>\n",
       "      <td>A</td>\n",
       "      <td>A15C A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>A15_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC</td>\n",
       "      <td>A16C</td>\n",
       "      <td>A</td>\n",
       "      <td>A16C A</td>\n",
       "      <td>No centre of mass</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>A16_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AC</td>\n",
       "      <td>A53C</td>\n",
       "      <td>A</td>\n",
       "      <td>A53C A</td>\n",
       "      <td>35.218287</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>A53_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AC</td>\n",
       "      <td>A55C</td>\n",
       "      <td>A</td>\n",
       "      <td>A55C A</td>\n",
       "      <td>31.791704</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>A55_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AC</td>\n",
       "      <td>A60C</td>\n",
       "      <td>A</td>\n",
       "      <td>A60C A</td>\n",
       "      <td>38.926044</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>A60_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28115</th>\n",
       "      <td>YW</td>\n",
       "      <td>Y597W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y597W B</td>\n",
       "      <td>40.613433</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Y597_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28116</th>\n",
       "      <td>YW</td>\n",
       "      <td>Y608W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y608W B</td>\n",
       "      <td>33.528173</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Y608_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28117</th>\n",
       "      <td>YW</td>\n",
       "      <td>Y638W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y638W B</td>\n",
       "      <td>46.007273</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Y638_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28118</th>\n",
       "      <td>YW</td>\n",
       "      <td>Y678W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y678W B</td>\n",
       "      <td>48.498844</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Y678_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28119</th>\n",
       "      <td>YW</td>\n",
       "      <td>Y711W</td>\n",
       "      <td>B</td>\n",
       "      <td>Y711W B</td>\n",
       "      <td>50.160048</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Y711_B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28120 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Mutation KatG mutations Chain Mutation Unique ID Distance from Site 1  \\\n",
       "0           AC           A15C     A             A15C A    No centre of mass   \n",
       "1           AC           A16C     A             A16C A    No centre of mass   \n",
       "2           AC           A53C     A             A53C A            35.218287   \n",
       "3           AC           A55C     A             A55C A            31.791704   \n",
       "4           AC           A60C     A             A60C A            38.926044   \n",
       "...        ...            ...   ...                ...                  ...   \n",
       "28115       YW          Y597W     B            Y597W B            40.613433   \n",
       "28116       YW          Y608W     B            Y608W B            33.528173   \n",
       "28117       YW          Y638W     B            Y638W B            46.007273   \n",
       "28118       YW          Y678W     B            Y678W B            48.498844   \n",
       "28119       YW          Y711W     B            Y711W B            50.160048   \n",
       "\n",
       "       d_volume  d_MW  d_hydropathy  d_Pi Start residue End residue  \\\n",
       "0         -19.9 -32.0          -0.7  0.93             A           C   \n",
       "1         -19.9 -32.0          -0.7  0.93             A           C   \n",
       "2         -19.9 -32.0          -0.7  0.93             A           C   \n",
       "3         -19.9 -32.0          -0.7  0.93             A           C   \n",
       "4         -19.9 -32.0          -0.7  0.93             A           C   \n",
       "...         ...   ...           ...   ...           ...         ...   \n",
       "28115     -34.2 -23.0          -0.4 -0.23             Y           W   \n",
       "28116     -34.2 -23.0          -0.4 -0.23             Y           W   \n",
       "28117     -34.2 -23.0          -0.4 -0.23             Y           W   \n",
       "28118     -34.2 -23.0          -0.4 -0.23             Y           W   \n",
       "28119     -34.2 -23.0          -0.4 -0.23             Y           W   \n",
       "\n",
       "       (Number of substitutions, min)  (Number of substitutions, max)  \\\n",
       "0                                   2                               3   \n",
       "1                                   2                               3   \n",
       "2                                   2                               3   \n",
       "3                                   2                               3   \n",
       "4                                   2                               3   \n",
       "...                               ...                             ...   \n",
       "28115                               2                               2   \n",
       "28116                               2                               2   \n",
       "28117                               2                               2   \n",
       "28118                               2                               2   \n",
       "28119                               2                               2   \n",
       "\n",
       "       (Number of substitutions, count) Residue Unique ID  \n",
       "0                                     8             A15_A  \n",
       "1                                     8             A16_A  \n",
       "2                                     8             A53_A  \n",
       "3                                     8             A55_A  \n",
       "4                                     8             A60_A  \n",
       "...                                 ...               ...  \n",
       "28115                                 2            Y597_B  \n",
       "28116                                 2            Y608_B  \n",
       "28117                                 2            Y638_B  \n",
       "28118                                 2            Y678_B  \n",
       "28119                                 2            Y711_B  \n",
       "\n",
       "[28120 rows x 15 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KatG_mutations_df.reset_index(inplace = True)\n",
    "del KatG_mutations_df['index']\n",
    "KatG_mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "9f92faec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adds depth column df to main df\n",
    "KatG_mutations_df.set_index('Residue Unique ID', inplace = True)\n",
    "distance_merge.set_index('Residue Unique ID', inplace = True)\n",
    "KatG_mutations_df = KatG_mutations_df.join(distance_merge)\n",
    "KatG_mutations_df.reset_index(inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "701244b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\user\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3444: DtypeWarning: Columns (6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,304,305,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363,364,365,366,367,368,369,370,371,372,373,374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,404,405,406,407,408,409,410,411,412,413,414,415,416,417,418,419,420,421,422,423,424,425,426,427,428,429,430,431,432,433,434,435,436,437,438,439,440,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455,456,457,458,459,460,461,462,463,464,465,466,467,468,469,470,471,472,473,474,475,476,477,478,479,480,481,482,483,484,485,486,487,488,489,490,491,492,493,494,495,496,497,498,499,500,501,502,503,504,505,506,507,508,509,510,511,512,513,514,515,516,517,518,519,520,521,522,523,524,525,526,527,528,529,530,531,532,533,534,535,536,537,538,539,540,541,542,543,544,545,546,547,548,549,550,551,552,553,554,555,556,557,558,559,560,561,562,563,564,565,566,567,568,569,570,571,572,573,574,575,576,577,578,579,580,581,582,583,584,585,586,587,588,589,590,591,592,593,594,595,596,597,598,599,600,601,602,603,604,605,606,607,608,609,610,611,612,613,614,615,616,617,618,619,620,621,622,623,624,625,626,627,628,629,630,631,632,633,634,635,636,637,638,639,640,641,642,643,644,645,646,647,648,649,650,651,652,653,654,655,656,657,658,659,660,661,662,663,664,665,666,667,668,669,670,671,672,673,674,675,676,677,678,679,680,681,682,683,684,685,686,687,688,689,690,691,692,693,694,695,696,697,698,699,700,701,702,703,704,705,706,707,708,709,710,711,712,713,714,715,716,717,718,719,720) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "#loads in DSSP csv file for KatG secondary structure \n",
    "KatG_ss = pd.read_csv(r'C:\\Users\\user\\anaconda3\\envs\\mdaenv\\Inh_resistance\\KatG_DSSP_ss.csv')\n",
    "KatG_ss = KatG_ss.iloc[:,:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "229b3288",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adds unique ID to ss dataframe\n",
    "ss_df_unique_ID = []\n",
    "ss_df_row_no = -1\n",
    "for i in KatG_ss['Res_number']:\n",
    "    ss_df_row_no += 1\n",
    "    Res = KatG_ss.loc[ss_df_row_no,'Res']\n",
    "    Num = KatG_ss.loc[ss_df_row_no,'Res_number']\n",
    "    Chain = KatG_ss.loc[ss_df_row_no,'Chain']\n",
    "    Unique = str(Res+str(Num)+'_'+Chain)\n",
    "    ss_df_unique_ID.append(Unique)\n",
    "KatG_ss['Residue_Unique_ID'] = ss_df_unique_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "e387e1d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gives each DSSP category a new column for a binary readout if the residue fits into that ss\n",
    "List = ['H','O','T','S','G','E','B','I']\n",
    "Dict = {}\n",
    "\n",
    "for i in List:\n",
    "    Dict[i] = []\n",
    "    \n",
    "for i in KatG_ss['SS']:\n",
    "    for j in List:\n",
    "        if i == j:\n",
    "            Dict[j].append('1')\n",
    "        else:\n",
    "            Dict[j].append('0')\n",
    "            \n",
    "for i in List:\n",
    "    KatG_ss[i] = Dict[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "d2ba54dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# deletes columns to ensure no overlap with main characteristic df \n",
    "KatG_ss.drop(['Chain','Res_number','Res'], inplace = True, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "ea7d30d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "KatG_ss.set_index('Residue_Unique_ID', inplace = True)\n",
    "KatG_mutations_df.set_index('Residue Unique ID', inplace = True)\n",
    "KatG_mutations_df = KatG_mutations_df.join(KatG_ss)\n",
    "KatG_mutations_df.reset_index(inplace = True)\n",
    "KatG_mutations_df.rename(columns = {'index':'Residue Unique ID'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "2d2d2d04",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\user\\AppData\\Local\\Temp/ipykernel_19516/3423963753.py:10: RuntimeWarning: Mean of empty slice.\n",
      "  res_tf = KatG.select_atoms('resid '+ str(i) +' and segid A').tempfactors.mean()\n",
      "C:\\Users\\user\\anaconda3\\lib\\site-packages\\numpy\\core\\_methods.py:188: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ret = ret.dtype.type(ret / rcount)\n",
      "C:\\Users\\user\\AppData\\Local\\Temp/ipykernel_19516/3423963753.py:15: RuntimeWarning: Mean of empty slice.\n",
      "  res_tf = KatG.select_atoms('resid '+ str(i) +' and segid B').tempfactors.mean()\n"
     ]
    }
   ],
   "source": [
    "unique_id = []\n",
    "tf = []\n",
    "\n",
    "for i in range(1,741):\n",
    "    Res = KatG_seq[i-1]\n",
    "    Num = i \n",
    "    #append residue info in chain A\n",
    "    Unique = str(Res+str(Num)+'_A')\n",
    "    unique_id.append(Unique)\n",
    "    res_tf = KatG.select_atoms('resid '+ str(i) +' and segid A').tempfactors.mean()\n",
    "    tf.append(res_tf)\n",
    "    #append residue info in chain B\n",
    "    Unique = str(Res+str(Num)+'_B')\n",
    "    unique_id.append(Unique)\n",
    "    res_tf = KatG.select_atoms('resid '+ str(i) +' and segid B').tempfactors.mean()\n",
    "    tf.append(res_tf)\n",
    "    \n",
    "data = {'Residue_Unique_ID':unique_id, 'Tempfactor':tf}\n",
    "tf_df = pd.DataFrame (data, columns = ['Residue_Unique_ID','Tempfactor'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "d366a208",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_df.set_index('Residue_Unique_ID', inplace = True)\n",
    "KatG_mutations_df.set_index('Residue Unique ID', inplace = True)\n",
    "KatG_mutations_df = KatG_mutations_df.join(tf_df)\n",
    "KatG_mutations_df.reset_index(inplace = True)\n",
    "KatG_mutations_df.rename(columns = {'index':'Residue Unique ID'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "3adc7e05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# distance from heam "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "4ad8af8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Site1_A = KatG.select_atoms('resid 315 and segid A', 'resid 137 and segid A')\n",
    "Site1_A_CoM = Site1_A.center_of_mass()\n",
    "haem_distances_A = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "c04c9f75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([137.71549016, 110.062874  , 149.22819452])"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KatG.select_atoms('resid 315 and segid A').center_of_mass()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "61b3026e",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_id = []\n",
    "Heam_dist = []\n",
    "Hem_A_CoM = Hem_A.center_of_mass()\n",
    "Hem_B_CoM = Hem_B.center_of_mass()\n",
    "\n",
    "for i in range(1,741):\n",
    "    Res = KatG_seq[i-1]\n",
    "    Num = i \n",
    "    #append residue info in chain A\n",
    "    Unique = str(Res+str(Num)+'_A')\n",
    "    unique_id.append(Unique)\n",
    "    res_CoM = KatG.select_atoms('resid '+ str(i) +' and segid A').center_of_mass()\n",
    "    dist = numpy.linalg.norm(res_CoM-Hem_A_CoM)\n",
    "    Heam_dist.append(dist)\n",
    "    #append residue info in chain B\n",
    "    Unique = str(Res+str(Num)+'_B')\n",
    "    unique_id.append(Unique)\n",
    "    res_CoM = KatG.select_atoms('resid '+ str(i) +' and segid B').center_of_mass()\n",
    "    dist = numpy.linalg.norm(res_CoM-Hem_B_CoM)    \n",
    "    Heam_dist.append(dist)\n",
    "    \n",
    "data = {'Residue_Unique_ID':unique_id, 'Hem_dist':Heam_dist}\n",
    "Heam_dist_df = pd.DataFrame (data, columns = ['Residue_Unique_ID','Hem_dist'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d0bbe079",
   "metadata": {},
   "outputs": [],
   "source": [
    "Heam_dist_df.set_index('Residue_Unique_ID', inplace = True)\n",
    "KatG_mutations_df.set_index('Residue Unique ID', inplace = True)\n",
    "KatG_mutations_df = KatG_mutations_df.join(Heam_dist_df)\n",
    "KatG_mutations_df.reset_index(inplace = True)\n",
    "KatG_mutations_df.rename(columns = {'index':'Residue Unique ID'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "0c195a0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Site2_Distance - only in chain A as chain A is only used for features in ML model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "c27cccd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_id = []\n",
    "Site2_A_dist = []\n",
    "Site2_A_CoM = [130, 105, 151]\n",
    "# derived from KatG structure paper\n",
    "\n",
    "for i in range(1,741):\n",
    "    Res = KatG_seq[i-1]\n",
    "    Num = i \n",
    "    #append residue info in chain A\n",
    "    Unique = str(Res+str(Num)+'_A')\n",
    "    unique_id.append(Unique)\n",
    "    res_CoM = KatG.select_atoms('resid '+ str(i) +' and segid A').center_of_mass()\n",
    "    dist = numpy.linalg.norm(res_CoM-Site2_A_CoM)\n",
    "    Site2_A_dist.append(dist)\n",
    "    #append residue info in chain B\n",
    "    Unique = str(Res+str(Num)+'_B')\n",
    "    unique_id.append(Unique)\n",
    "    res_CoM = KatG.select_atoms('resid '+ str(i) +' and segid B').center_of_mass()\n",
    "    dist = numpy.linalg.norm(res_CoM-Site2_A_CoM)    \n",
    "    Site2_A_dist.append(dist)\n",
    "    \n",
    "data = {'Residue_Unique_ID':unique_id, 'Site2_dist':Site2_A_dist}\n",
    "Site2_dist_df = pd.DataFrame (data, columns = ['Residue_Unique_ID','Site2_dist'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "e2446ab4",
   "metadata": {},
   "outputs": [],
   "source": [
    "Site2_dist_df.set_index('Residue_Unique_ID', inplace = True)\n",
    "KatG_mutations_df.set_index('Residue Unique ID', inplace = True)\n",
    "KatG_mutations_df = KatG_mutations_df.join(Site2_dist_df)\n",
    "KatG_mutations_df.reset_index(inplace = True)\n",
    "KatG_mutations_df.rename(columns = {'index':'Residue Unique ID'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "ea4c812f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use SNAP2 to determine link between genotype and phentoype of given mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "aeae45f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "SNAP2_pd = pd.read_csv('SNAP2_values.csv')\n",
    "SNAP2_pd.drop(['Predicted Effect','Expected Accuracy'], inplace = True, axis = 1)\n",
    "SNAP2_pd.set_index('Variant', inplace = True)\n",
    "SNAP2_pd.rename(columns = {'Score':'SNAP2_score'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "4e9100c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNAP2_score</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M1A</th>\n",
       "      <td>-83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1R</th>\n",
       "      <td>-85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1N</th>\n",
       "      <td>-84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1D</th>\n",
       "      <td>-65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1C</th>\n",
       "      <td>-93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R740S</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R740T</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R740W</th>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R740Y</th>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R740V</th>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14800 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SNAP2_score\n",
       "Variant             \n",
       "M1A              -83\n",
       "M1R              -85\n",
       "M1N              -84\n",
       "M1D              -65\n",
       "M1C              -93\n",
       "...              ...\n",
       "R740S             14\n",
       "R740T              3\n",
       "R740W             77\n",
       "R740Y             40\n",
       "R740V             48\n",
       "\n",
       "[14800 rows x 1 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNAP2_pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "82ec3d75",
   "metadata": {},
   "outputs": [],
   "source": [
    "KatG_mutations_df.set_index('KatG mutations', inplace = True)\n",
    "KatG_mutations_df = KatG_mutations_df.join(SNAP2_pd)\n",
    "KatG_mutations_df.reset_index(inplace = True)\n",
    "KatG_mutations_df.rename(columns = {'index':'KatG mutations'}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "c2df0afc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KatG mutations</th>\n",
       "      <th>Residue Unique ID</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Chain</th>\n",
       "      <th>Mutation Unique ID</th>\n",
       "      <th>Distance from Site 1</th>\n",
       "      <th>d_volume</th>\n",
       "      <th>d_MW</th>\n",
       "      <th>d_hydropathy</th>\n",
       "      <th>d_Pi</th>\n",
       "      <th>Start residue</th>\n",
       "      <th>End residue</th>\n",
       "      <th>(Number of substitutions, min)</th>\n",
       "      <th>(Number of substitutions, max)</th>\n",
       "      <th>(Number of substitutions, count)</th>\n",
       "      <th>Depth</th>\n",
       "      <th>SS</th>\n",
       "      <th>H</th>\n",
       "      <th>O</th>\n",
       "      <th>T</th>\n",
       "      <th>S</th>\n",
       "      <th>G</th>\n",
       "      <th>E</th>\n",
       "      <th>B</th>\n",
       "      <th>I</th>\n",
       "      <th>Tempfactor</th>\n",
       "      <th>Hem_dist</th>\n",
       "      <th>Site2_dist</th>\n",
       "      <th>SNAP2_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A106C</td>\n",
       "      <td>A106_A</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>A106C A</td>\n",
       "      <td>14.629585</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>13.382314</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>82.300003</td>\n",
       "      <td>10.589587</td>\n",
       "      <td>14.275680</td>\n",
       "      <td>-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A106C</td>\n",
       "      <td>A106_B</td>\n",
       "      <td>AC</td>\n",
       "      <td>B</td>\n",
       "      <td>A106C B</td>\n",
       "      <td>14.465213</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>0.93</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>10.619754</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>91.110001</td>\n",
       "      <td>10.373444</td>\n",
       "      <td>62.981121</td>\n",
       "      <td>-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A106D</td>\n",
       "      <td>A106_A</td>\n",
       "      <td>AD</td>\n",
       "      <td>A</td>\n",
       "      <td>A106D A</td>\n",
       "      <td>14.629585</td>\n",
       "      <td>-22.5</td>\n",
       "      <td>-44.0</td>\n",
       "      <td>5.3</td>\n",
       "      <td>3.23</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>13.382314</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>82.300003</td>\n",
       "      <td>10.589587</td>\n",
       "      <td>14.275680</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A106D</td>\n",
       "      <td>A106_B</td>\n",
       "      <td>AD</td>\n",
       "      <td>B</td>\n",
       "      <td>A106D B</td>\n",
       "      <td>14.465213</td>\n",
       "      <td>-22.5</td>\n",
       "      <td>-44.0</td>\n",
       "      <td>5.3</td>\n",
       "      <td>3.23</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>10.619754</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>91.110001</td>\n",
       "      <td>10.373444</td>\n",
       "      <td>62.981121</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A106E</td>\n",
       "      <td>A106_A</td>\n",
       "      <td>AE</td>\n",
       "      <td>A</td>\n",
       "      <td>A106E A</td>\n",
       "      <td>14.629585</td>\n",
       "      <td>-49.8</td>\n",
       "      <td>-58.0</td>\n",
       "      <td>5.3</td>\n",
       "      <td>2.78</td>\n",
       "      <td>A</td>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>13.382314</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>82.300003</td>\n",
       "      <td>10.589587</td>\n",
       "      <td>14.275680</td>\n",
       "      <td>-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28115</th>\n",
       "      <td>Y98T</td>\n",
       "      <td>Y98_B</td>\n",
       "      <td>YT</td>\n",
       "      <td>B</td>\n",
       "      <td>Y98T B</td>\n",
       "      <td>17.788426</td>\n",
       "      <td>77.5</td>\n",
       "      <td>62.1</td>\n",
       "      <td>-0.6</td>\n",
       "      <td>0.06</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>4.926100</td>\n",
       "      <td>O</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>93.639999</td>\n",
       "      <td>15.165136</td>\n",
       "      <td>71.318767</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28116</th>\n",
       "      <td>Y98V</td>\n",
       "      <td>Y98_A</td>\n",
       "      <td>YV</td>\n",
       "      <td>A</td>\n",
       "      <td>Y98V A</td>\n",
       "      <td>17.955051</td>\n",
       "      <td>53.6</td>\n",
       "      <td>64.1</td>\n",
       "      <td>-5.5</td>\n",
       "      <td>-0.30</td>\n",
       "      <td>Y</td>\n",
       "      <td>V</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>6.643434</td>\n",
       "      <td>O</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88.239998</td>\n",
       "      <td>15.225818</td>\n",
       "      <td>8.741972</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28117</th>\n",
       "      <td>Y98V</td>\n",
       "      <td>Y98_B</td>\n",
       "      <td>YV</td>\n",
       "      <td>B</td>\n",
       "      <td>Y98V B</td>\n",
       "      <td>17.788426</td>\n",
       "      <td>53.6</td>\n",
       "      <td>64.1</td>\n",
       "      <td>-5.5</td>\n",
       "      <td>-0.30</td>\n",
       "      <td>Y</td>\n",
       "      <td>V</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>4.926100</td>\n",
       "      <td>O</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>93.639999</td>\n",
       "      <td>15.165136</td>\n",
       "      <td>71.318767</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28118</th>\n",
       "      <td>Y98W</td>\n",
       "      <td>Y98_A</td>\n",
       "      <td>YW</td>\n",
       "      <td>A</td>\n",
       "      <td>Y98W A</td>\n",
       "      <td>17.955051</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6.643434</td>\n",
       "      <td>O</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88.239998</td>\n",
       "      <td>15.225818</td>\n",
       "      <td>8.741972</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28119</th>\n",
       "      <td>Y98W</td>\n",
       "      <td>Y98_B</td>\n",
       "      <td>YW</td>\n",
       "      <td>B</td>\n",
       "      <td>Y98W B</td>\n",
       "      <td>17.788426</td>\n",
       "      <td>-34.2</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>Y</td>\n",
       "      <td>W</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4.926100</td>\n",
       "      <td>O</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>93.639999</td>\n",
       "      <td>15.165136</td>\n",
       "      <td>71.318767</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28120 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      KatG mutations Residue Unique ID Mutation Chain Mutation Unique ID  \\\n",
       "0              A106C            A106_A       AC     A            A106C A   \n",
       "1              A106C            A106_B       AC     B            A106C B   \n",
       "2              A106D            A106_A       AD     A            A106D A   \n",
       "3              A106D            A106_B       AD     B            A106D B   \n",
       "4              A106E            A106_A       AE     A            A106E A   \n",
       "...              ...               ...      ...   ...                ...   \n",
       "28115           Y98T             Y98_B       YT     B             Y98T B   \n",
       "28116           Y98V             Y98_A       YV     A             Y98V A   \n",
       "28117           Y98V             Y98_B       YV     B             Y98V B   \n",
       "28118           Y98W             Y98_A       YW     A             Y98W A   \n",
       "28119           Y98W             Y98_B       YW     B             Y98W B   \n",
       "\n",
       "      Distance from Site 1  d_volume  d_MW  d_hydropathy  d_Pi Start residue  \\\n",
       "0                14.629585     -19.9 -32.0          -0.7  0.93             A   \n",
       "1                14.465213     -19.9 -32.0          -0.7  0.93             A   \n",
       "2                14.629585     -22.5 -44.0           5.3  3.23             A   \n",
       "3                14.465213     -22.5 -44.0           5.3  3.23             A   \n",
       "4                14.629585     -49.8 -58.0           5.3  2.78             A   \n",
       "...                    ...       ...   ...           ...   ...           ...   \n",
       "28115            17.788426      77.5  62.1          -0.6  0.06             Y   \n",
       "28116            17.955051      53.6  64.1          -5.5 -0.30             Y   \n",
       "28117            17.788426      53.6  64.1          -5.5 -0.30             Y   \n",
       "28118            17.955051     -34.2 -23.0          -0.4 -0.23             Y   \n",
       "28119            17.788426     -34.2 -23.0          -0.4 -0.23             Y   \n",
       "\n",
       "      End residue  (Number of substitutions, min)  \\\n",
       "0               C                               2   \n",
       "1               C                               2   \n",
       "2               D                               1   \n",
       "3               D                               1   \n",
       "4               E                               1   \n",
       "...           ...                             ...   \n",
       "28115           T                               2   \n",
       "28116           V                               2   \n",
       "28117           V                               2   \n",
       "28118           W                               2   \n",
       "28119           W                               2   \n",
       "\n",
       "       (Number of substitutions, max)  (Number of substitutions, count)  \\\n",
       "0                                   3                                 8   \n",
       "1                                   3                                 8   \n",
       "2                                   2                                 8   \n",
       "3                                   2                                 8   \n",
       "4                                   2                                 8   \n",
       "...                               ...                               ...   \n",
       "28115                               3                                 8   \n",
       "28116                               3                                 8   \n",
       "28117                               3                                 8   \n",
       "28118                               2                                 2   \n",
       "28119                               2                                 2   \n",
       "\n",
       "           Depth SS  H  O  T  S  G  E  B  I  Tempfactor   Hem_dist  \\\n",
       "0      13.382314  H  1  0  0  0  0  0  0  0   82.300003  10.589587   \n",
       "1      10.619754  H  1  0  0  0  0  0  0  0   91.110001  10.373444   \n",
       "2      13.382314  H  1  0  0  0  0  0  0  0   82.300003  10.589587   \n",
       "3      10.619754  H  1  0  0  0  0  0  0  0   91.110001  10.373444   \n",
       "4      13.382314  H  1  0  0  0  0  0  0  0   82.300003  10.589587   \n",
       "...          ... .. .. .. .. .. .. .. .. ..         ...        ...   \n",
       "28115   4.926100  O  0  1  0  0  0  0  0  0   93.639999  15.165136   \n",
       "28116   6.643434  O  0  1  0  0  0  0  0  0   88.239998  15.225818   \n",
       "28117   4.926100  O  0  1  0  0  0  0  0  0   93.639999  15.165136   \n",
       "28118   6.643434  O  0  1  0  0  0  0  0  0   88.239998  15.225818   \n",
       "28119   4.926100  O  0  1  0  0  0  0  0  0   93.639999  15.165136   \n",
       "\n",
       "       Site2_dist  SNAP2_score  \n",
       "0       14.275680          -27  \n",
       "1       62.981121          -27  \n",
       "2       14.275680            5  \n",
       "3       62.981121            5  \n",
       "4       14.275680           -7  \n",
       "...           ...          ...  \n",
       "28115   71.318767           41  \n",
       "28116    8.741972           34  \n",
       "28117   71.318767           34  \n",
       "28118    8.741972           22  \n",
       "28119   71.318767           22  \n",
       "\n",
       "[28120 rows x 29 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KatG_mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "0a6b060e",
   "metadata": {},
   "outputs": [],
   "source": [
    "KatG_mutations_df.to_csv('KatG_mutations_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "8f7ed465",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.385879743759846\n"
     ]
    }
   ],
   "source": [
    "#distance between site 1 and site 2\n",
    "Site1 =  [135.55687286, 113.46202071, 148.68011475]\n",
    "Site2 = [130, 105, 151]\n",
    "x = numpy.asarray(Site1)\n",
    "y = numpy.asarray(Site2)\n",
    "dist = numpy.linalg.norm(x-y)\n",
    "print(dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59bfad9d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
